
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.3">
    
    
      
        <title>pyHLAMSA</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.50c56a3b.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="css/mkdocstrings.css">
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#pyhlamsa-python-utility-for-star-alleles-sequence-alignment" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="pyHLAMSA" class="md-header__button md-logo" aria-label="pyHLAMSA" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            pyHLAMSA
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              pyHLAMSA: Python Utility for Star Alleles Sequence Alignment
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/linnil1/pyHLAMSA" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="pyHLAMSA" class="md-nav__button md-logo" aria-label="pyHLAMSA" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    pyHLAMSA
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/linnil1/pyHLAMSA" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    pyHLAMSA: Python Utility for Star Alleles Sequence Alignment
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    pyHLAMSA: Python Utility for Star Alleles Sequence Alignment
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#run-in-command-line" class="md-nav__link">
    <span class="md-ellipsis">
      Run in command line
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    <span class="md-ellipsis">
      Features
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-read-sequences-from-database-in-one-line" class="md-nav__link">
    <span class="md-ellipsis">
      1. Read sequences from database in one line
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-merge-gene-and-nuc-msa-is-quiet-simple" class="md-nav__link">
    <span class="md-ellipsis">
      2. Merge gene and nuc MSA is quiet simple
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-block-based-selection-you-can-select-any-intronexon" class="md-nav__link">
    <span class="md-ellipsis">
      3. Block-based selection: You can select ANY intron/exon.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-easy-to-compare-alleles-select-and-print" class="md-nav__link">
    <span class="md-ellipsis">
      4. Easy to compare alleles: select and print
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-some-useful-operation" class="md-nav__link">
    <span class="md-ellipsis">
      5. Some useful operation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-write-and-export-the-msa" class="md-nav__link">
    <span class="md-ellipsis">
      6. Write and export the MSA
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#todo" class="md-nav__link">
    <span class="md-ellipsis">
      TODO
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirement" class="md-nav__link">
    <span class="md-ellipsis">
      Requirement
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-cyp" class="md-nav__link">
    <span class="md-ellipsis">
      Appendix: CYP
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup-document" class="md-nav__link">
    <span class="md-ellipsis">
      Setup Document
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#test" class="md-nav__link">
    <span class="md-ellipsis">
      Test
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Test">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tested-version" class="md-nav__link">
    <span class="md-ellipsis">
      Tested Version
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#some-qas" class="md-nav__link">
    <span class="md-ellipsis">
      Some QAs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citations" class="md-nav__link">
    <span class="md-ellipsis">
      Citations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#document" class="md-nav__link">
    <span class="md-ellipsis">
      Document
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa" class="md-nav__link">
    <span class="md-ellipsis">
      pyhlamsa.gene.genemsa
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa" class="md-nav__link">
    <span class="md-ellipsis">
      Genemsa
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Genemsa">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__add__" class="md-nav__link">
    <span class="md-ellipsis">
      __add__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      __getitem__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      __len__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      __repr__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.append" class="md-nav__link">
    <span class="md-ellipsis">
      append()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.assume_label" class="md-nav__link">
    <span class="md-ellipsis">
      assume_label()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.calculate_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_frequency()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.contains" class="md-nav__link">
    <span class="md-ellipsis">
      contains()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.copy" class="md-nav__link">
    <span class="md-ellipsis">
      copy()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.extend" class="md-nav__link">
    <span class="md-ellipsis">
      extend()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.fill_incomplete" class="md-nav__link">
    <span class="md-ellipsis">
      fill_incomplete()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.format_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      format_alignment()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get" class="md-nav__link">
    <span class="md-ellipsis">
      get()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_allele_or_error" class="md-nav__link">
    <span class="md-ellipsis">
      get_allele_or_error()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_block" class="md-nav__link">
    <span class="md-ellipsis">
      get_block()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_block_index" class="md-nav__link">
    <span class="md-ellipsis">
      get_block_index()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_block_interval" class="md-nav__link">
    <span class="md-ellipsis">
      get_block_interval()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_block_length" class="md-nav__link">
    <span class="md-ellipsis">
      get_block_length()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_cigar" class="md-nav__link">
    <span class="md-ellipsis">
      get_cigar()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_consensus" class="md-nav__link">
    <span class="md-ellipsis">
      get_consensus()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_length" class="md-nav__link">
    <span class="md-ellipsis">
      get_length()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_reference" class="md-nav__link">
    <span class="md-ellipsis">
      get_reference()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_sequence_names" class="md-nav__link">
    <span class="md-ellipsis">
      get_sequence_names()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.get_variantion_base" class="md-nav__link">
    <span class="md-ellipsis">
      get_variantion_base()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.items" class="md-nav__link">
    <span class="md-ellipsis">
      items()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.list_alleles" class="md-nav__link">
    <span class="md-ellipsis">
      list_alleles()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.list_blocks" class="md-nav__link">
    <span class="md-ellipsis">
      list_blocks()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.list_index" class="md-nav__link">
    <span class="md-ellipsis">
      list_index()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.merge_exon" class="md-nav__link">
    <span class="md-ellipsis">
      merge_exon()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.meta_to_json" class="md-nav__link">
    <span class="md-ellipsis">
      meta_to_json()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.print_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      print_alignment()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.print_alignment_diff" class="md-nav__link">
    <span class="md-ellipsis">
      print_alignment_diff()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.print_alignment_from_center" class="md-nav__link">
    <span class="md-ellipsis">
      print_alignment_from_center()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.print_snv" class="md-nav__link">
    <span class="md-ellipsis">
      print_snv()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.remove_allele" class="md-nav__link">
    <span class="md-ellipsis">
      remove_allele()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.reset_index" class="md-nav__link">
    <span class="md-ellipsis">
      reset_index()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.reverse_complement" class="md-nav__link">
    <span class="md-ellipsis">
      reverse_complement()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.save_msa" class="md-nav__link">
    <span class="md-ellipsis">
      save_msa()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.select_allele" class="md-nav__link">
    <span class="md-ellipsis">
      select_allele()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.select_block" class="md-nav__link">
    <span class="md-ellipsis">
      select_block()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.select_complete" class="md-nav__link">
    <span class="md-ellipsis">
      select_complete()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.select_exon" class="md-nav__link">
    <span class="md-ellipsis">
      select_exon()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.select_incomplete" class="md-nav__link">
    <span class="md-ellipsis">
      select_incomplete()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.set_blocks" class="md-nav__link">
    <span class="md-ellipsis">
      set_blocks()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.set_reference" class="md-nav__link">
    <span class="md-ellipsis">
      set_reference()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.shrink" class="md-nav__link">
    <span class="md-ellipsis">
      shrink()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.size" class="md-nav__link">
    <span class="md-ellipsis">
      size()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.sort" class="md-nav__link">
    <span class="md-ellipsis">
      sort()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.sort_name" class="md-nav__link">
    <span class="md-ellipsis">
      sort_name()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.split_block" class="md-nav__link">
    <span class="md-ellipsis">
      split_block()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_MultipleSeqAlignment" class="md-nav__link">
    <span class="md-ellipsis">
      to_MultipleSeqAlignment()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_bam" class="md-nav__link">
    <span class="md-ellipsis">
      to_bam()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      to_fasta()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_gff" class="md-nav__link">
    <span class="md-ellipsis">
      to_gff()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_imgt_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      to_imgt_alignment()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_records" class="md-nav__link">
    <span class="md-ellipsis">
      to_records()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.to_vcf" class="md-nav__link">
    <span class="md-ellipsis">
      to_vcf()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene.genemsa.Genemsa.truth" class="md-nav__link">
    <span class="md-ellipsis">
      truth()
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla" class="md-nav__link">
    <span class="md-ellipsis">
      pyhlamsa.gene_family.hla
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa" class="md-nav__link">
    <span class="md-ellipsis">
      HLAmsa
    </span>
  </a>
  
    <nav class="md-nav" aria-label="HLAmsa">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      __getitem__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.items" class="md-nav__link">
    <span class="md-ellipsis">
      items()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.list_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      list_db_gene()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.list_genes" class="md-nav__link">
    <span class="md-ellipsis">
      list_genes()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla.HLAmsa.read_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      read_db_gene()
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex" class="md-nav__link">
    <span class="md-ellipsis">
      pyhlamsa.gene_family.hla_ex
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX" class="md-nav__link">
    <span class="md-ellipsis">
      HLAmsaEX
    </span>
  </a>
  
    <nav class="md-nav" aria-label="HLAmsaEX">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      __getitem__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.items" class="md-nav__link">
    <span class="md-ellipsis">
      items()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.list_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      list_db_gene()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.list_genes" class="md-nav__link">
    <span class="md-ellipsis">
      list_genes()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.hla_ex.HLAmsaEX.read_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      read_db_gene()
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir" class="md-nav__link">
    <span class="md-ellipsis">
      pyhlamsa.gene_family.kir
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa" class="md-nav__link">
    <span class="md-ellipsis">
      KIRmsa
    </span>
  </a>
  
    <nav class="md-nav" aria-label="KIRmsa">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      __getitem__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.items" class="md-nav__link">
    <span class="md-ellipsis">
      items()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.list_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      list_db_gene()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.list_genes" class="md-nav__link">
    <span class="md-ellipsis">
      list_genes()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.kir.KIRmsa.read_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      read_db_gene()
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp" class="md-nav__link">
    <span class="md-ellipsis">
      pyhlamsa.gene_family.cyp
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa" class="md-nav__link">
    <span class="md-ellipsis">
      CYPmsa
    </span>
  </a>
  
    <nav class="md-nav" aria-label="CYPmsa">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      __getitem__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.items" class="md-nav__link">
    <span class="md-ellipsis">
      items()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.list_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      list_db_gene()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.list_genes" class="md-nav__link">
    <span class="md-ellipsis">
      list_genes()
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pyhlamsa.gene_family.cyp.CYPmsa.read_db_gene" class="md-nav__link">
    <span class="md-ellipsis">
      read_db_gene()
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="pyhlamsa-python-utility-for-star-alleles-sequence-alignment">pyHLAMSA: Python Utility for Star Alleles Sequence Alignment</h1>
<p>pyHLAMSA is a Python tool for handling Multiple Sequence Alignments (MSA) of genes with star alleles,
specifically focusing on IMGT-HLA, IMGT-KIR, and CYP datasets. Key features include:</p>
<ul>
<li><strong>MSA Download:</strong> Fetch MSAs for latest/specific versions.</li>
<li><strong>Gene and Allele Management:</strong> Add, delete, list, and select genes and alleles.</li>
<li><strong>Intron/Exon Operations:</strong> Manipulate specific intron/exon segments.</li>
<li><strong>Position and Region Selection:</strong> Choose targeted positions or regions.</li>
<li><strong>Alignment Operations:</strong> Add, view, concatenate, crop, and find differences.</li>
<li><strong>Genomic and Nucleotide Merging:</strong> Seamlessly merge sequences.</li>
<li><strong>Variant Operations:</strong> Calculate consensus, gather statistics, and perform variant-related tasks.</li>
<li><strong>Biopython Compatibility:</strong> Transform alignments into Biopython MultipleSeqAlignment.</li>
<li><strong>Format Conversion:</strong> Load and save MSAs in various formats (VCF, MSF, TXT, BAM, GFF).</li>
</ul>
<p>pyHLAMSA streamlines complex gene sequence analysis, offering efficient tools for researchers.</p>
<h2 id="run-in-command-line">Run in command line</h2>
<p>You can simply use this package by command line.</p>
<p>If you want to use more powerful function, try the APIs written in below sections.</p>
<div class="highlight"><pre><span></span><code>pip3<span class="w"> </span>install<span class="w"> </span>git+https://github.com/linnil1/pyHLAMSA
<span class="c1"># show help</span>
pyhlamsa<span class="w"> </span>-h
<span class="c1"># download kir</span>
pyhlamsa<span class="w"> </span>download<span class="w"> </span>--family<span class="w"> </span>kir<span class="w"> </span>--db-folder<span class="w"> </span>tmpdir/tmp_kir_db<span class="w"> </span>--version<span class="w"> </span><span class="m">2100</span><span class="w">  </span>tmpdir/kir<span class="w"> </span>--include-genes<span class="w"> </span>KIR2DL1<span class="w"> </span>KIR2DL2
<span class="c1"># view the msa</span>
pyhlamsa<span class="w"> </span>view<span class="w"> </span>tmpdir/kir.KIR2DL1<span class="w"> </span>--position<span class="w"> </span><span class="m">3</span>-100<span class="w"> </span>--include-alleles<span class="w"> </span>KIR2DL1*consensus<span class="w"> </span>KIR2DL1*063
<span class="c1"># save the intron1+exon1 region to kir1.*</span>
pyhlamsa<span class="w"> </span>view<span class="w"> </span>tmpdir/kir.KIR2DL1<span class="w"> </span>--region<span class="w"> </span>intron1<span class="w"> </span>exon1<span class="w"> </span>--name<span class="w"> </span>tmpdir/kir1<span class="w"> </span>--save<span class="w"> </span>--bam<span class="w"> </span>--gff<span class="w"> </span>--vcf<span class="w"> </span>--fasta-gapless<span class="w"> </span>--fasta-msa
</code></pre></div>
<h2 id="features">Features</h2>
<h3 id="1-read-sequences-from-database-in-one-line">1. Read sequences from database in one line</h3>
<p><code>HLAmsa</code> provided a simple way to read HLA data</p>
<p>It can automatically download and read the sequences</p>
<div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">pyhlamsa</span> <span class="kn">import</span> <span class="n">HLAmsa</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">hla</span> <span class="o">=</span> <span class="n">HLAmsa</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">filetype</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span>
                 <span class="n">version</span><span class="o">=</span><span class="s2">&quot;3470&quot;</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">hla</span><span class="o">.</span><span class="n">list_genes</span><span class="p">())</span>
<span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">]</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">hla</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">4101</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">301</span><span class="p">)</span> <span class="n">exon1</span><span class="p">(</span><span class="mi">80</span><span class="p">)</span> <span class="n">intron1</span><span class="p">(</span><span class="mi">130</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">273</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span> <span class="n">intron3</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">287</span><span class="p">)</span> <span class="n">intron4</span><span class="p">(</span><span class="mi">119</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">117</span><span class="p">)</span> <span class="n">intron5</span><span class="p">(</span><span class="mi">444</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">33</span><span class="p">)</span> <span class="n">intron6</span><span class="p">(</span><span class="mi">143</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">48</span><span class="p">)</span> <span class="n">intron7</span><span class="p">(</span><span class="mi">170</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">302</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
<p><code>KIRmsa</code> can also read sequences of KIR</p>
<div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">pyhlamsa</span> <span class="kn">import</span> <span class="n">KIRmsa</span>

<span class="c1"># If don&#39;t specific the genes, it will read all genes.</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">kir</span> <span class="o">=</span> <span class="n">KIRmsa</span><span class="p">(</span><span class="n">ipd_folder</span><span class="o">=</span><span class="s2">&quot;KIR_v2100&quot;</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s2">&quot;2100&quot;</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">kir</span><span class="o">.</span><span class="n">list_genes</span><span class="p">())</span>
<span class="p">[</span><span class="s1">&#39;KIR2DL1&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DL2&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DL3&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DL4&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DL5&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DP1&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DS1&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DS2&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DS3&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DS4&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR2DS5&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR3DL1&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR3DL2&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR3DL3&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR3DP1&#39;</span><span class="p">,</span> <span class="s1">&#39;KIR3DS1&#39;</span><span class="p">]</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">kir</span><span class="p">[</span><span class="s2">&quot;KIR2DL1&quot;</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">KIR2DL1</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">173</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">268</span><span class="p">)</span> <span class="n">exon1</span><span class="p">(</span><span class="mi">34</span><span class="p">)</span> <span class="n">intron1</span><span class="p">(</span><span class="mi">964</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">36</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">728</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">282</span><span class="p">)</span> <span class="n">intron3</span><span class="p">(</span><span class="mi">1441</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">300</span><span class="p">)</span> <span class="n">intron4</span><span class="p">(</span><span class="mi">1534</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">294</span><span class="p">)</span> <span class="n">intron5</span><span class="p">(</span><span class="mi">3157</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">51</span><span class="p">)</span> <span class="n">intron6</span><span class="p">(</span><span class="mi">4270</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">102</span><span class="p">)</span> <span class="n">intron7</span><span class="p">(</span><span class="mi">462</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">53</span><span class="p">)</span> <span class="n">intron8</span><span class="p">(</span><span class="mi">98</span><span class="p">)</span> <span class="n">exon9</span><span class="p">(</span><span class="mi">177</span><span class="p">)</span> <span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">510</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
<p>Kind note: In our modules, <code>exon</code> is actually CDS, so it doesn't include <code>UTR</code>.</p>
<h3 id="2-merge-gene-and-nuc-msa-is-quiet-simple">2. Merge gene and nuc MSA is quiet simple</h3>
<p>This main features give us a chance to use genomic MSA and nucleotide MSA at the same time.</p>
<p>The nucleotide MSA is a exon-only sequence, thus we fill the intron with <code>E</code> after merged.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># merge gen and nuc sequences when loading</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">hla</span> <span class="o">=</span> <span class="n">HLAmsa</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">],</span> <span class="n">filetype</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
                 <span class="n">imgt_alignment_folder</span><span class="o">=</span><span class="s2">&quot;alignment_v3470&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">hla</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7349</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">301</span><span class="p">)</span> <span class="n">exon1</span><span class="p">(</span><span class="mi">80</span><span class="p">)</span> <span class="n">intron1</span><span class="p">(</span><span class="mi">130</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">273</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">436</span><span class="p">)</span> <span class="n">intron3</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">361</span><span class="p">)</span> <span class="n">intron4</span><span class="p">(</span><span class="mi">119</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">117</span><span class="p">)</span> <span class="n">intron5</span><span class="p">(</span><span class="mi">444</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">33</span><span class="p">)</span> <span class="n">intron6</span><span class="p">(</span><span class="mi">143</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">48</span><span class="p">)</span> <span class="n">intron7</span><span class="p">(</span><span class="mi">170</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">302</span><span class="p">)</span><span class="o">&gt;</span>


<span class="c1"># or manually</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span> <span class="o">=</span> <span class="n">HLAmsa</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">filetype</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> 
<span class="o">&gt;&gt;&gt;</span>                <span class="n">imgt_alignment_folder</span><span class="o">=</span><span class="s2">&quot;alignment_v3470&quot;</span><span class="p">)[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a_gen</span><span class="p">)</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">4101</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">301</span><span class="p">)</span> <span class="n">exon1</span><span class="p">(</span><span class="mi">80</span><span class="p">)</span> <span class="n">intron1</span><span class="p">(</span><span class="mi">130</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">273</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span> <span class="n">intron3</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">287</span><span class="p">)</span> <span class="n">intron4</span><span class="p">(</span><span class="mi">119</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">117</span><span class="p">)</span> <span class="n">intron5</span><span class="p">(</span><span class="mi">444</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">33</span><span class="p">)</span> <span class="n">intron6</span><span class="p">(</span><span class="mi">143</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">48</span><span class="p">)</span> <span class="n">intron7</span><span class="p">(</span><span class="mi">170</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">302</span><span class="p">)</span><span class="o">&gt;</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">a_nuc</span> <span class="o">=</span> <span class="n">HLAmsa</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">filetype</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">,</span> 
<span class="o">&gt;&gt;&gt;</span>                <span class="n">imgt_alignment_folder</span><span class="o">=</span><span class="s2">&quot;alignment_v3470&quot;</span><span class="p">)[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a_nuc</span><span class="p">)</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7353</span> <span class="n">block</span><span class="o">=</span><span class="n">exon1</span><span class="p">(</span><span class="mi">80</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">436</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">361</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">117</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">33</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">48</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span><span class="o">&gt;</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span> <span class="o">=</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;A*03:437Q&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">a_nuc</span><span class="p">))</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7349</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">301</span><span class="p">)</span> <span class="n">exon1</span><span class="p">(</span><span class="mi">80</span><span class="p">)</span> <span class="n">intron1</span><span class="p">(</span><span class="mi">130</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">273</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">436</span><span class="p">)</span> <span class="n">intron3</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span> <span class="n">exon4</span><span class="p">(</span><span class="mi">361</span><span class="p">)</span> <span class="n">intron4</span><span class="p">(</span><span class="mi">119</span><span class="p">)</span> <span class="n">exon5</span><span class="p">(</span><span class="mi">117</span><span class="p">)</span> <span class="n">intron5</span><span class="p">(</span><span class="mi">444</span><span class="p">)</span> <span class="n">exon6</span><span class="p">(</span><span class="mi">33</span><span class="p">)</span> <span class="n">intron6</span><span class="p">(</span><span class="mi">143</span><span class="p">)</span> <span class="n">exon7</span><span class="p">(</span><span class="mi">48</span><span class="p">)</span> <span class="n">intron7</span><span class="p">(</span><span class="mi">170</span><span class="p">)</span> <span class="n">exon8</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">302</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
<h3 id="3-block-based-selection-you-can-select-any-intronexon">3. Block-based selection: You can select ANY intron/exon.</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># select exon2 and exon3</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23</span> <span class="o">=</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>  <span class="c1"># 1-base</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">exon23</span><span class="p">)</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">4100</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span><span class="o">&gt;</span>


<span class="c1"># select exon2 + intron2 + exon3</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">e2i2e3</span> <span class="o">=</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">])</span>  <span class="c1"># 0-base</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">e2i2e3</span><span class="p">)</span>
<span class="o">&lt;</span><span class="n">A</span>  <span class="n">alleles</span><span class="o">=</span><span class="mi">4100</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">intron2</span><span class="p">(</span><span class="mi">273</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
<h3 id="4-easy-to-compare-alleles-select-and-print">4. Easy to compare alleles: select and print</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># select first 10 alleles</span>
<span class="o">&gt;&gt;</span> <span class="n">exon23_10</span> <span class="o">=</span> <span class="n">exon23</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="n">exon23</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">()[:</span><span class="mi">10</span><span class="p">])</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">exon23_10</span><span class="p">)</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">10</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span><span class="o">&gt;</span>

<span class="c1"># print it</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23_10</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">()</span>
                  <span class="mi">812</span>                                                      <span class="mi">1136</span>
                    <span class="o">|</span>                                                         <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>     <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">03</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">04</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">05</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">06</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">07</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">08</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">09</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">10</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">-</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">--</span> <span class="o">----------</span>


<span class="c1"># using regex to select</span>
<span class="c1"># &quot;|&quot; indicate the border of block, in this case, it&#39;s the border of exon2 and exon3</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23_1field</span> <span class="o">=</span> <span class="n">exon23</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;A\*.*:01:01:01$&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23_1field</span><span class="o">.</span><span class="n">print_alignment_diff</span><span class="p">()</span>
                  <span class="mi">812</span>                                                      <span class="mi">1136</span>
                    <span class="o">|</span>                                                         <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="n">ACCGAGCGAA</span> <span class="n">CCTGGGGACC</span> <span class="n">CTGCGCGGCT</span> <span class="n">ACTACAACCA</span> <span class="n">GAGCGAGGAC</span> <span class="n">G</span><span class="o">|</span> <span class="n">GTTCTCACA</span> <span class="n">CC</span><span class="o">.</span><span class="n">ATCCAGA</span> <span class="n">TAATGTATGG</span> <span class="n">CTGCGACG</span><span class="o">..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.</span><span class="n">G</span><span class="o">------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">----------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">----------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">23</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">A</span><span class="o">---</span> <span class="o">----</span><span class="n">C</span><span class="o">---</span><span class="n">T</span><span class="o">-</span> <span class="n">GC</span><span class="o">--</span><span class="n">T</span><span class="o">-</span><span class="n">C</span><span class="o">---</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.</span><span class="n">C</span><span class="o">------</span> <span class="o">-</span><span class="n">G</span><span class="o">----</span><span class="n">T</span><span class="o">---</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">25</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">A</span><span class="o">--</span><span class="n">G</span> <span class="o">----</span><span class="n">C</span><span class="o">---</span><span class="n">T</span><span class="o">-</span> <span class="n">GC</span><span class="o">--</span><span class="n">T</span><span class="o">-</span><span class="n">C</span><span class="o">---</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">29</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">-</span><span class="n">G</span><span class="o">--------</span> <span class="o">----</span><span class="n">C</span><span class="o">---..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">30</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">----------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">32</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">A</span><span class="o">--</span><span class="n">G</span> <span class="o">----</span><span class="n">C</span><span class="o">---</span><span class="n">T</span><span class="o">-</span> <span class="n">GC</span><span class="o">--</span><span class="n">T</span><span class="o">-</span><span class="n">C</span><span class="o">---</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">-</span><span class="n">G</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">-</span><span class="n">G</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">36</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">----------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">66</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">68</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">-</span><span class="n">G</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">69</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.</span><span class="n">G</span><span class="o">------</span> <span class="n">GG</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">74</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">------</span><span class="n">T</span><span class="o">-</span><span class="n">G</span><span class="o">-</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">--------</span><span class="n">C</span><span class="o">-</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">-</span><span class="n">G</span><span class="o">--------</span> <span class="o">--------..</span> <span class="o">..........</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">80</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">-|</span> <span class="o">---------</span> <span class="o">--.-------</span> <span class="o">----------</span> <span class="o">--------..</span> <span class="o">..........</span>


<span class="c1"># show only variantiation</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23_1field</span><span class="o">.</span><span class="n">print_snv</span><span class="p">()</span>
<span class="n">Total</span> <span class="n">variantion</span><span class="p">:</span> <span class="mi">71</span>
                         <span class="mi">536</span> <span class="mi">537</span>  <span class="mi">541</span>          <span class="mi">565</span> <span class="mi">567</span> <span class="mi">570</span>          <span class="mi">593</span>          <span class="mi">640</span>          <span class="mi">654</span> <span class="mi">658</span>          <span class="mi">684</span>          <span class="mi">721</span>      
                           <span class="o">|</span>   <span class="o">|</span>    <span class="o">|</span>            <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span>   <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span>      
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="n">ATTTCT</span>   <span class="n">TCAC</span> <span class="n">ATCCG</span><span class="o">|</span> <span class="n">CCGGC</span> <span class="n">CG</span>  <span class="n">CGG</span> <span class="n">GGA</span><span class="o">.</span><span class="n">G</span> <span class="o">|</span> <span class="n">ATCGCCGTGG</span> <span class="o">|</span> <span class="o">.</span><span class="n">G</span><span class="o">.</span><span class="n">ACACG</span><span class="o">.</span><span class="n">C</span> <span class="o">|</span> <span class="n">CGTGCGGTTC</span> <span class="n">GACA</span><span class="o">|</span> <span class="n">AGA</span><span class="o">..</span><span class="n">A</span> <span class="n">GATG</span><span class="o">|</span> <span class="n">CGGGCG</span> <span class="n">CCGT</span><span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">23</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">C</span><span class="o">---</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">25</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">29</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">-----</span><span class="n">A</span>   <span class="n">C</span><span class="o">---</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">---------</span><span class="n">T</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">-----</span><span class="n">A</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">30</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">C</span><span class="o">---</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="n">A</span><span class="o">-</span>  <span class="n">T</span><span class="o">--</span> <span class="n">A</span><span class="o">--.-</span> <span class="o">|</span> <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">32</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">---------</span><span class="n">T</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">-----</span><span class="n">A</span>   <span class="n">C</span><span class="o">---</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">36</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..-</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">66</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">68</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">69</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="n">A</span><span class="o">---</span> <span class="n">C</span><span class="o">----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">74</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">.-.-----.-</span> <span class="o">|</span> <span class="o">---------</span><span class="n">T</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">80</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="o">|</span> <span class="o">------</span>   <span class="o">----</span> <span class="o">-----|</span> <span class="o">-----</span> <span class="o">--</span>  <span class="o">---</span> <span class="o">---.-</span> <span class="o">|</span> <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">|</span> <span class="o">.-.--</span><span class="n">T</span><span class="o">--.-</span> <span class="o">|</span> <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">----|</span> <span class="o">---..</span><span class="n">G</span> <span class="o">----|</span> <span class="o">------</span> <span class="o">----|</span>
</code></pre></div>
<h3 id="5-some-useful-operation">5. Some useful operation</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Calculate variant frequency of ( A T C G - ) per base</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">exon23</span><span class="o">.</span><span class="n">calculate_frequency</span><span class="p">()[:</span><span class="mi">10</span><span class="p">])</span>
<span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4095</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4097</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4098</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4095</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">911</span><span class="p">,</span> <span class="mi">3188</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4097</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4099</span><span class="p">],</span> <span class="p">[</span><span class="mi">4097</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4090</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4097</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>


<span class="c1"># Get consensus(The largest one in frequency) among the msa</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23</span><span class="o">.</span><span class="n">get_consensus</span><span class="p">(</span><span class="n">include_gap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="s1">&#39;GCTCCC-ACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGGCCGCGGGGA----GCCCCGCTTCATCGCCGTGGGC-----------------------TACGTGGACG-ACACG-CAGTTCGTGCGGTTCGACAGCGACGCCGCGAGCCAGAGGATGGAGCCG--------------------CGGGCGCCGTGGATA-GAGCAGGAGGGGCCGGAGTATTGGGACCAGGAGACACGGA-------------A-TGTGAAGGCCCACTCACAGACTGACCGAGTGGACCTGGGGACCCTGCGCGGCTACTACAACCAGAGCGAGGCCGGTTCTCACACC-ATCCAGATGATGTATGGCTGCGACG--------------TGGGG-TCGGACGGGCGCTTCCTCCGCGGGTACCAGCA---GGACGCCTACGACGGCAAGGATTAC---ATCGCCCTGAAC------------------------GAGGACCTGCGCTCTTGGACCGCGGCGGAC--------ATGGCGGCTCAGATCACCAAGCGC-AAGT----GGGAGG--CGGCCC-ATGT------------------------------------------GGCGG-AGCAGTTGAGAGCCTACCTGGAGGGCACG--------TGCGTG----GAGTGGCTCCG--CAGATA-CCTGGAGAACGGGAAGGAGACGCTGCAGC-----------------GCACGG&#39;</span>


<span class="c1"># Add sequence into MSA</span>
<span class="c1"># include_gap=False in get_consensus will ignore the frequency of gap. i.e. choose one of ATCG</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a_merged</span> <span class="o">=</span> <span class="n">hla</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">consensus_seq</span> <span class="o">=</span> <span class="n">a_merged</span><span class="o">.</span><span class="n">get_consensus</span><span class="p">(</span><span class="n">include_gap</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a_merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;A*consensus&quot;</span><span class="p">,</span> <span class="n">consensus_seq</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a_merged</span><span class="o">.</span><span class="n">fill_imcomplete</span><span class="p">(</span><span class="s2">&quot;A*consensus&quot;</span><span class="p">)</span>


<span class="c1"># Shrink: remove gap if all bases in the column are gap</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">exon23_10</span><span class="o">.</span><span class="n">shrink</span><span class="p">()</span><span class="o">.</span><span class="n">print_snv</span><span class="p">()</span>
 <span class="n">gDNA</span>               <span class="mi">200</span>
                    <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="n">AAGGCCCACT</span> <span class="n">CACAGACTGA</span> <span class="n">CCGAGCGAAC</span> <span class="n">CTGGGGACCC</span> <span class="n">TGCGCGGCTA</span> <span class="n">CTACAACCAG</span> <span class="n">AGCGAGGACG</span><span class="o">|</span> <span class="n">GTTCTCACAC</span> <span class="n">CATCCAGATA</span> <span class="n">ATGTATGGCT</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>     <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">03</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">04</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">05</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">06</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">07</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">08</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">09</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">10</span>      <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------|</span> <span class="o">----------</span> <span class="o">----------</span> <span class="o">----------</span>


<span class="c1"># select specific column </span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="p">[</span><span class="mi">12</span><span class="p">:</span><span class="mi">100</span><span class="p">]</span>
<span class="o">&lt;</span><span class="n">A</span>  <span class="n">alleles</span><span class="o">=</span><span class="mi">4100</span> <span class="n">block</span><span class="o">=</span><span class="p">(</span><span class="mi">88</span><span class="p">)</span><span class="o">&gt;</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="p">[[</span><span class="mi">12</span><span class="p">,</span><span class="mi">100</span><span class="p">]]</span><span class="o">.</span><span class="n">print_alignment_diff</span><span class="p">()</span>
 <span class="n">gDNA</span>               <span class="mi">0</span>
                    <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="n">GG</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>     <span class="o">-</span><span class="n">G</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">03</span>      <span class="n">GG</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">04</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">05</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">06</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">07</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">08</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">09</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">10</span>      <span class="o">--</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">11</span>      <span class="n">GG</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">12</span>      <span class="n">GG</span>


<span class="c1"># concat</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">2</span><span class="p">])</span> <span class="o">+</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">3</span><span class="p">]))</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">4100</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
<h3 id="6-write-and-export-the-msa">6. Write and export the MSA</h3>
<p><strong>Causion: If you run <code>merge_exon</code>, or the msa is from <code>filetype=['gen', 'nuc']</code>,
You should fill the <code>E</code> BEFORE save it.</strong></p>
<p>You can fill it by consensus_seq shown before.</p>
<ul>
<li>
<p>MultipleSeqAlignment</p>
<p>Transfer to <a href="https://biopython.org/docs/1.75/api/Bio.Align.html#Bio.Align.MultipleSeqAlignment">MultipleSeqAlignment</a></p>
<div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a_gen</span><span class="o">.</span><span class="n">to_MultipleSeqAlignment</span><span class="p">())</span>
<span class="n">Alignment</span> <span class="k">with</span> <span class="mi">4100</span> <span class="n">rows</span> <span class="ow">and</span> <span class="mi">3866</span> <span class="n">columns</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">03</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">04</span>
<span class="o">--------------------------------------------...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">05</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">06</span>
<span class="o">--------------------------------------------...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">07</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">08</span>
<span class="o">--------------------------------------------...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">09</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">10</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">11</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">12</span>
<span class="o">--------------------------------------------...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">13</span>
<span class="o">--------------------------------------------...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">14</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">15</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">16</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">17</span>
<span class="n">CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTG</span><span class="o">...</span><span class="n">AAA</span> <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">18</span>
<span class="o">...</span>
<span class="o">--------------------------------------------...---</span> <span class="n">A</span><span class="o">*</span><span class="mi">80</span><span class="p">:</span><span class="mi">07</span>
</code></pre></div>
</li>
<li>
<p>list of SeqRecord
    <div class="highlight"><pre><span></span><code><span class="c1"># Save as MSA</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">a_gen</span><span class="o">.</span><span class="n">to_records</span><span class="p">(</span><span class="n">gap</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="s2">&quot;filename.msa.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="c1"># Save as no-gapped sequences</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">a_gen</span><span class="o">.</span><span class="n">to_records</span><span class="p">(</span><span class="n">gap</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="s2">&quot;filename.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>fasta
    <div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="s2">&quot;filename.msa.fa&quot;</span><span class="p">,</span> <span class="n">gap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">a_gen</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="s2">&quot;filename.fa&quot;</span><span class="p">,</span> <span class="n">gap</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>bam
    <div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">to_bam</span><span class="p">(</span><span class="s2">&quot;filename.bam&quot;</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>gff
    <div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">to_gff</span><span class="p">(</span><span class="s2">&quot;filename.gff&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p>After save the MSA as bam and gff, you can show the alignments on IGV
<img alt="msa_igv_example" src="https://raw.githubusercontent.com/linnil1/pyHLAMSA/main/HLA_msa.png" /></p>
</li>
<li>
<p>vcf
    <div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">to_vcf</span><span class="p">(</span><span class="s2">&quot;filename.vcf.gz&quot;</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>IMGT MSA format (xx_gen.txt)
    <div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">to_imgt_alignment</span><span class="p">(</span><span class="s2">&quot;A_gen.txt&quot;</span><span class="p">)</span>
<span class="n">a_gen</span><span class="o">.</span><span class="n">to_imgt_alignment</span><span class="p">(</span><span class="s2">&quot;A_nuc.txt&quot;</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>save/load</p>
<p>Save our model in fasta and json, where json contains block, index information</p>
<div class="highlight"><pre><span></span><code><span class="n">a_gen</span><span class="o">.</span><span class="n">save_msa</span><span class="p">(</span><span class="s2">&quot;a_gen.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;a_gen.json&quot;</span><span class="p">)</span>
<span class="n">a_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">load_msa</span><span class="p">(</span><span class="s2">&quot;a_gen.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;a_gen.json&quot;</span><span class="p">)</span>
</code></pre></div>
</li>
<li>
<p>load msa from other format</p>
<p>pyHLAMSA only support reading from <code>MultipleSeqAlignment</code>,
which is very useful object,
can be generate by reading MSA by <code>Bio.AlignIO</code>.</p>
<p>Checkout <a href="https://biopython.org/wiki/AlignIO#file-formats">https://biopython.org/wiki/AlignIO#file-formats</a> for format supporting.</p>
<p>For example
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pyhlamsa</span> <span class="kn">import</span> <span class="n">Genemsa</span>
<span class="n">msa</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">from_MultipleSeqAlignment</span><span class="p">(</span><span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">your_data_path</span><span class="p">,</span> <span class="n">your_data_format</span><span class="p">))</span>
</code></pre></div></p>
</li>
</ul>
<h2 id="todo">TODO</h2>
<ul>
<li>[x] Testing</li>
<li>[x] Main function</li>
<li>[x] exon-only sequence handling</li>
<li>[x] Reading from file</li>
<li>[x] Some useful function: <code>copy</code>, <code>remove</code>, <code>get_sequence_names</code>, <code>__len__</code>, <code>size</code>, <code>split</code>, <code>concat</code></li>
<li>[x] Cannot handle pseudo exon</li>
<li>[x] merge blocks and labels</li>
<li>[x] Fix KIR when merge gen and nuc</li>
<li>[x] Use index to trace orignal position</li>
<li>[x] CYP</li>
<li>[x] Set reference</li>
<li>[x] Download latest version of IMGT or IPD</li>
<li>[x] Remove seqtype</li>
<li>[x] Split code</li>
<li>[x] save to VCF</li>
<li>[x] Add command line usage</li>
<li>[x] Remove msaio</li>
<li>[x] Rewrite gene Mixins (Too complicated)</li>
<li>[x] Fix selction and removeing allele by regex</li>
<li>[x] Change to ACGT order</li>
<li>[x] Rename: split -&gt; split_block, remove -&gt; remove_allele</li>
<li>[ ] CDS != exon, (rename it?)</li>
<li>[ ] Rename: pyHLAMSA -&gt; py_star_msa (MAYBE, becuase the project is originally written for HLA)</li>
</ul>
<h2 id="requirement">Requirement</h2>
<ul>
<li>python3.9</li>
<li>biopython</li>
<li>pysam</li>
<li>wget</li>
<li>git</li>
</ul>
<h2 id="installation">Installation</h2>
<div class="highlight"><pre><span></span><code>pip3<span class="w"> </span>install<span class="w"> </span>git+https://github.com/linnil1/pyHLAMSA
<span class="c1"># or</span>
git<span class="w"> </span>clone<span class="w"> </span>https://github.com/linnil1/pyHLAMSA
pip3<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>pyHLAMSA
</code></pre></div>
<h2 id="appendix-cyp">Appendix: CYP</h2>
<p>Steps:</p>
<ol>
<li>Download fasta from <a href="https://www.pharmvar.org/download">https://www.pharmvar.org/download</a></li>
<li>unzip to <code>./pharmvar-5.1.10</code></li>
<li>Read it by pyhlamsa
<div class="highlight"><pre><span></span><code><span class="c1"># 4. Read it</span>
<span class="kn">from</span> <span class="nn">pyhlamsa</span> <span class="kn">import</span> <span class="n">CYPmsa</span>
<span class="n">cyp</span> <span class="o">=</span> <span class="n">CYPmsa</span><span class="p">(</span><span class="n">pharmvar_folder</span><span class="o">=</span><span class="s2">&quot;./pharmvar-5.1.10&quot;</span><span class="p">)</span>

<span class="c1"># 5. Test it</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">cyp</span><span class="p">[</span><span class="s1">&#39;CYP26A1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">format_variantion_base</span><span class="p">())</span>

                       <span class="mi">6407</span>         <span class="mi">6448</span>         <span class="mi">8142</span>
                          <span class="o">|</span>            <span class="o">|</span>            <span class="o">|</span>
 <span class="n">CYP26A1</span><span class="o">*</span><span class="mf">1.001</span>       <span class="n">GCGAGCGCGG</span> <span class="o">|</span> <span class="n">ATGTTCCGAA</span> <span class="o">|</span> <span class="n">TCGGGTGTGT</span>
 <span class="n">CYP26A1</span><span class="o">*</span><span class="mf">3.001</span>       <span class="o">----------</span> <span class="o">|</span> <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">|</span> <span class="o">----------</span>
 <span class="n">CYP26A1</span><span class="o">*</span><span class="mf">2.001</span>       <span class="o">-----</span><span class="n">A</span><span class="o">----</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">----------</span>
 <span class="n">CYP26A1</span><span class="o">*</span><span class="mf">4.001</span>       <span class="o">----------</span> <span class="o">|</span> <span class="o">----------</span> <span class="o">|</span> <span class="o">-----</span><span class="n">C</span><span class="o">----</span>
</code></pre></div></li>
</ol>
<h2 id="setup-document">Setup Document</h2>
<div class="highlight"><pre><span></span><code>pip3<span class="w"> </span>install<span class="w"> </span>mkdocs-material<span class="w"> </span>mkdocstrings<span class="o">[</span>python-lagacy<span class="o">]==</span><span class="m">0</span>.18
mkdocs<span class="w"> </span>serve
</code></pre></div>
<p>I use python-lagacy 0.18 because <code>inherited_members</code> is not support now</p>
<h2 id="test">Test</h2>
<div class="highlight"><pre><span></span><code>pip3<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
pip3<span class="w"> </span>install<span class="w"> </span>pytest<span class="w"> </span>black<span class="w"> </span>mypy
pytest
black<span class="w"> </span>pyhlamsa
mypy<span class="w"> </span>pyhlamsa
</code></pre></div>
<h3 id="tested-version">Tested Version</h3>
<ul>
<li>CYP: 5.2.2</li>
<li>HLA: 3.49.0</li>
<li>KIR: 2.10.0 (The DB in 2.11.0 contains bugs in KIR2DL4/5)</li>
</ul>
<h2 id="some-qas">Some QAs</h2>
<blockquote>
<p>Why not inherit Bio.AlignIO.MultipleSeqAlignment?</p>
</blockquote>
<p>The class does't support lot of functions than I expected.</p>
<p>And it's tidious to overwrite most of the functions to maintain our blocks information</p>
<blockquote>
<p>Why not use numpy to save the sequence?</p>
</blockquote>
<p>Performance issue is not my bottle-neck yet.</p>
<h2 id="citations">Citations</h2>
<ul>
<li>HLA
    Robinson J, Barker DJ, Georgiou X, Cooper MA, Flicek P, Marsh SGE: IPD-IMGT/HLA Database. Nucleic Acids Research (2020), 48:D948-55</li>
<li>IMGT
    Robinson J, Barker DJ, Georgiou X, Cooper MA, Flicek P, Marsh SGE
    IPD-IMGT/HLA Database
    Nucleic Acids Research (2020) 48:D948-55</li>
<li>CYP(PharmVar)
    Pharmacogene Variation Consortium (PharmVar) at www.PharmVar.org using the following references: Gaedigk et al. 2018, CPT 103(3):399-401 (PMID 29134625); Gaedigk et al. 2020, CPT 107(1):43-46 (PMID 31758698) and Gaedigk et al. 2021, CPT 110(3):542-545 (PMID 34091888)</li>
<li>This github</li>
</ul>
<h2 id="document">Document</h2>
<p>See <a href="See https://linnil1.github.io/pyHLAMSA">https://linnil1.github.io/pyHLAMSA</a></p>


  <div class="doc doc-object doc-module">

<a id="pyhlamsa.gene.genemsa"></a>
    <div class="doc doc-contents first">

      <p>Gene Module.</p>
<p>All the msa operation are merged into Genemsa</p>



  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h2 id="pyhlamsa.gene.genemsa.Genemsa" class="doc doc-heading">
        <code>
Genemsa            (<span title="pyhlamsa.gene.io_operation.GenemsaIO">GenemsaIO</span>, <span title="pyhlamsa.gene.text_operation.GenemsaTextOp">GenemsaTextOp</span>, <span title="pyhlamsa.gene.base.GenemsaBase">GenemsaBase</span>)
        </code>



</h2>

    <div class="doc doc-contents ">

      <p>Genemsa: core object in pyhlamsa</p>
<p>More details written in GenemsaBase</p>
<p>Basic usage:
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pyhlamsa</span> <span class="kn">import</span> <span class="n">Genemsa</span>
<span class="n">msa</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="p">(</span><span class="s2">&quot;test1&quot;</span><span class="p">)</span>
<span class="n">msa</span><span class="o">.</span><span class="n">set_blocks</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">7</span><span class="p">])</span>
<span class="n">msa</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;other&quot;</span><span class="p">)</span>
<span class="n">msa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;GGAGGA--CA&quot;</span><span class="p">)</span>
<span class="n">msa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;GGGGGAA--A&quot;</span><span class="p">)</span>
<span class="n">msa</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">()</span>
                  <span class="mi">1</span>   <span class="mi">4</span>     <span class="mi">9</span>
                  <span class="o">|</span>   <span class="o">|</span>     <span class="o">|</span>
<span class="n">A</span>                 <span class="n">GGA</span><span class="o">|</span><span class="n">GGA</span><span class="o">--|</span><span class="n">CA</span>
<span class="n">B</span>                 <span class="n">GGG</span><span class="o">|</span><span class="n">GGAA</span><span class="o">-|-</span><span class="n">A</span>
</code></pre></div></p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Genemsa</span><span class="p">(</span>
    <span class="n">GenemsaIO</span><span class="p">,</span>
    <span class="n">GenemsaTextOp</span><span class="p">,</span>
    <span class="n">GenemsaBase</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Genemsa: core object in pyhlamsa</span>

<span class="sd">    More details written in GenemsaBase</span>

<span class="sd">    Basic usage:</span>
<span class="sd">    ``` python</span>
<span class="sd">    from pyhlamsa import Genemsa</span>
<span class="sd">    msa = Genemsa(&quot;test1&quot;)</span>
<span class="sd">    msa.set_blocks([3,7])</span>
<span class="sd">    msa.assume_label(&quot;other&quot;)</span>
<span class="sd">    msa.append(&quot;A&quot;, &quot;GGAGGA--CA&quot;)</span>
<span class="sd">    msa.append(&quot;B&quot;, &quot;GGGGGAA--A&quot;)</span>
<span class="sd">    msa.print_alignment()</span>
<span class="sd">                      1   4     9</span>
<span class="sd">                      |   |     |</span>
<span class="sd">    A                 GGA|GGA--|CA</span>
<span class="sd">    B                 GGG|GGAA-|-A</span>
<span class="sd">    ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__add__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Concat 2 MSA</span>

<span class="sd">    Example:</span>
<span class="sd">      &gt;&gt;&gt; print(a_gen.select_exon([2]) + a_gen.select_exon([3]))</span>
<span class="sd">      &lt;A nuc alleles=4100 block=exon2(335) exon3(413)&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">names0</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
    <span class="n">names1</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
    <span class="k">if</span> <span class="n">names0</span> <span class="o">!=</span> <span class="n">names1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Can not concat because some allele is miss: &quot;</span>
            <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">names0</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">names1</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">blocks</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">blocks</span><span class="p">))</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">msa</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="n">seq</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__getitem__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="n">Any</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="n">Any</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the region of the sequences by index (start from 0),</span>
<span class="sd">    but block information will not preserved</span>

<span class="sd">    Example:</span>
<span class="sd">      &gt;&gt;&gt; msa = Genemsa(&quot;A&quot;, &quot;gen&quot;)</span>
<span class="sd">      &gt;&gt;&gt; msa.read_alignment_file(&quot;A_gen.txt&quot;)</span>
<span class="sd">      &gt;&gt;&gt; # Inspect 50-100bp in the MSA</span>
<span class="sd">      &gt;&gt;&gt; extract_msa = msa[50:100]</span>
<span class="sd">      &gt;&gt;&gt; print(extract_msa)</span>

<span class="sd">      &gt;&gt;&gt; # Inspect 2nd 3rd 5th bp in the MSA</span>
<span class="sd">      &gt;&gt;&gt; extract_msa = msa[[1,2,4]]</span>
<span class="sd">      &gt;&gt;&gt; print(extract_msa)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">index</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;MSA is empty&quot;</span><span class="p">)</span>

    <span class="c1"># Extract specific region in alignment</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
        <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">slice</span><span class="p">):</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="p">{</span><span class="n">allele</span><span class="p">:</span> <span class="n">seq</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="nb">list</span><span class="p">)):</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">allele</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index</span><span class="p">])</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index</span><span class="p">]</span>
    <span class="c1"># Fail</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Bad usage&quot;</span><span class="p">)</span>

    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">set_blocks</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">alleles</span><span class="o">.</span><span class="n">values</span><span class="p">())))])</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;Unamed&#39;</span><span class="p">,</span> <span class="n">alleles</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span> <span class="n">blocks</span><span class="p">:</span> <span class="n">Iterable</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">index</span><span class="p">:</span> <span class="n">Iterable</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">reference</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Unamed&quot;</span><span class="p">,</span>
    <span class="n">alleles</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">blocks</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">BlockInfo</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">index</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">IndexInfo</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">reference</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Attributes:</span>
<span class="sd">        gene_name: The name of the gene</span>

<span class="sd">        alleles: MSA data.</span>

<span class="sd">            Allele name as the key and the sequence string as the value.</span>


<span class="sd">        blocks: list of block information</span>
<span class="sd">        index: list of index(position) information</span>
<span class="sd">        reference: allele of the msa (Optional)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">blocks</span> <span class="ow">or</span> <span class="p">[]))</span>  <span class="c1"># intron exon length</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">index</span> <span class="ow">or</span> <span class="p">[]))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="n">reference</span>

    <span class="k">if</span> <span class="n">alleles</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__iter__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterator</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterator</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iter allele like iter(dict)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__len__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the number of alleles in the MSA</span>

<span class="sd">    Example:</span>
<span class="sd">      &gt;&gt;&gt; len(a_gen)</span>
<span class="sd">      4100</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.__repr__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Show name, number of alleles and block infos in this MSA&quot;&quot;&quot;</span>
    <span class="n">block_info</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">b</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">(</span><span class="si">{</span><span class="n">b</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()])</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&lt;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span><span class="si">}</span><span class="s2"> alleles=</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="si">}</span><span class="s2"> block=</span><span class="si">{</span><span class="n">block_info</span><span class="si">}</span><span class="s2">&gt;&quot;</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.append" class="doc doc-heading">
<code class="highlight language-python"><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Add a sequence into MSA (inplace)</p>
<p>Make sure the sequence length is same as in MSA</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">append</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add a sequence into MSA (inplace)</span>

<span class="sd">    Make sure the sequence length is same as in MSA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># OK to add 0 length string</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq</span>
        <span class="k">return</span> <span class="bp">self</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_length</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Length not match to alignments&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> already exist&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.assume_label" class="doc doc-heading">
<code class="highlight language-python"><span class="n">assume_label</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">seq_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;gen&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>It will automatically generate the block's label
according on <code>seq_type</code>. (inplace)</p>
<p>seq_type:
  * gen: 5UTR-exon1-intron1-exon2-...-exon9-3UTR
  * nuc: exon1-exon2-...-exon9
  * other: block1-block2-block3-...</p>
<p>!!! block_length
    If manually assign the block_length, the old block will be cleared.</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">assume_label</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">seq_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gen&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    It will automatically generate the block&#39;s label</span>
<span class="sd">    according on `seq_type`. (inplace)</span>

<span class="sd">    seq_type:</span>
<span class="sd">      * gen: 5UTR-exon1-intron1-exon2-...-exon9-3UTR</span>
<span class="sd">      * nuc: exon1-exon2-...-exon9</span>
<span class="sd">      * other: block1-block2-block3-...</span>

<span class="sd">    block_length:</span>
<span class="sd">        If manually assign the block_length, the old block will be cleared.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;This msa doesn&#39;t have any blocks&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">seq_type</span> <span class="o">==</span> <span class="s2">&quot;gen&quot;</span><span class="p">:</span>
        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">()</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="mi">3</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">blk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;five_prime_UTR&quot;</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;5UTR&quot;</span>
            <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;three_prime_UTR&quot;</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;3UTR&quot;</span>
            <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;exon&quot;</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;exon</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="mi">2</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;intron&quot;</span>
                <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;intron</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="mi">2</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">elif</span> <span class="n">seq_type</span> <span class="o">==</span> <span class="s2">&quot;nuc&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">blk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">):</span>
            <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;exon&quot;</span>
            <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;exon</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">blk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">):</span>
            <span class="n">blk</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;gene_fragment&quot;</span>
            <span class="n">blk</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;block</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="c1"># inplace to reset the index</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">index</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.calculate_frequency" class="doc doc-heading">
<code class="highlight language-python"><span class="n">calculate_frequency</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Calculate ATCG and gap frequency of each bp in MSA</p>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>frequency (list[list[int]])</code></td>
      <td><p>Each items contains the number of ATCG and gap.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">calculate_frequency</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate ATCG and gap frequency of each bp in MSA</span>

<span class="sd">    Returns:</span>
<span class="sd">      frequency (list[list[int]]):</span>
<span class="sd">        Each items contains the number of ATCG and gap.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="n">freqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">),</span> <span class="n">i</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">),</span> <span class="n">i</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">),</span> <span class="n">i</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">),</span> <span class="n">i</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)]</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">freqs</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.contains" class="doc doc-heading">
<code class="highlight language-python"><span class="n">contains</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Implement <code>in</code> operator</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">msa_a</span> <span class="o">=</span> <span class="n">HLAmsa</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="o">&gt;&gt;&gt;</span> <span class="s2">&quot;A*01:01:01:01&quot;</span> <span class="ow">in</span> <span class="n">msa_a</span>
<span class="kc">True</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">contains</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Implement `in` operator</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; msa_a = HLAmsa(&quot;A&quot;)[&quot;A&quot;]</span>
<span class="sd">        &gt;&gt;&gt; &quot;A*01:01:01:01&quot; in msa_a</span>
<span class="sd">        True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">allele</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.copy" class="doc doc-heading">
<code class="highlight language-python"><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">copy_allele</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Clone a new MSA.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>copy_allele</code></td>
        <td><code>bool</code></td>
        <td><p>Copy the sequences as well</p></td>
        <td><code>True</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">copy_allele</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Clone a new MSA.</span>

<span class="sd">    Args:</span>
<span class="sd">      copy_allele: Copy the sequences as well</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Child&#39;s Type</span>
    <span class="n">Genemsa</span> <span class="o">=</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span><span class="p">,</span>
        <span class="n">blocks</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
        <span class="n">reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span>
        <span class="n">alleles</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">alleles</span> <span class="k">if</span> <span class="n">copy_allele</span> <span class="k">else</span> <span class="p">{},</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.extend" class="doc doc-heading">
<code class="highlight language-python"><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Add MSA's alleles into this MSA (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">extend</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add MSA&#39;s alleles into this MSA (inplace)&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa</span><span class="p">,</span> <span class="n">GenemsaBase</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">[</span><span class="n">b</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()]</span> <span class="o">!=</span> <span class="p">[</span>
            <span class="n">b</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">msa</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()</span>
        <span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Block length is different&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">msa</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.fill_incomplete" class="doc doc-heading">
<code class="highlight language-python"><span class="n">fill_incomplete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Fill the <code>E</code> in exon-only sequences with ref_allele sequence (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">fill_incomplete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Fill the `E` in exon-only sequences with ref_allele sequence (inplace)&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ref_allele</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_allele</span><span class="si">}</span><span class="s2"> not found&quot;</span><span class="p">)</span>

    <span class="n">ref_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="s2">&quot;E&quot;</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
            <span class="c1"># replace it</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;E&quot;</span> <span class="k">else</span> <span class="n">ref_seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))]</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.format_alignment" class="doc doc-heading">
<code class="highlight language-python"><span class="n">format_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterator</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Transfer MSA to string(generator)
To the things in step3 and step4.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>kwargs_format</code></td>
        <td><code>Any</code></td>
        <td><ul>
<li>show_position_set(set[int]): Force to show the position</li>
<li>max_page_width(int): The max width of each line (Default = 140char)</li>
</ul></td>
        <td><code>{}</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">format_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterator</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transfer MSA to string(generator)</span>
<span class="sd">    To the things in step3 and step4.</span>

<span class="sd">    Args:</span>
<span class="sd">      kwargs_format:</span>
<span class="sd">        * show_position_set(set[int]): Force to show the position</span>
<span class="sd">        * max_page_width(int): The max width of each line (Default = 140char)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;MSA is empty&quot;</span><span class="p">)</span>
    <span class="n">pages</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_format_page</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">)</span>
    <span class="k">yield from</span> <span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_apply_page</span><span class="p">,</span> <span class="n">pages</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the sequence by allele name</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the sequence by allele name&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_allele_or_error" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_allele_or_error</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the sequence by allele name</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>allele</code></td>
        <td><code>str</code></td>
        <td><p>Allele name. If not provided, reference allele are used</p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>allele_name (str) and its sequence (str)</code></td>
      <td></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_allele_or_error</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the sequence by allele name</span>

<span class="sd">    Args:</span>
<span class="sd">      allele (str): Allele name. If not provided, reference allele are used</span>

<span class="sd">    Returns:</span>
<span class="sd">      allele_name (str) and its sequence (str):</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">allele</span><span class="p">:</span>
        <span class="n">allele</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_reference</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">allele</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">allele</span><span class="si">}</span><span class="s2"> not found&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">allele</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_block" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_block</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">BlockInfo</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">BlockInfo</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get block by str or id</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_block</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">BlockInput</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BlockInfo</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get block by str or id&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">get_block_index</span><span class="p">(</span><span class="n">block</span><span class="p">)]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_block_index" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_block_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">BlockInfo</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">int</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Find the index of the block</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_block_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">BlockInput</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the index of the block&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">block</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">block</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">i</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">block</span><span class="p">,</span> <span class="n">BlockInfo</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">i</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">block</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
        <span class="nb">id</span> <span class="o">=</span> <span class="n">block</span>
        <span class="k">if</span> <span class="nb">id</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">()</span> <span class="o">+</span> <span class="nb">id</span>
        <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="nb">id</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_length</span><span class="p">():</span>
            <span class="k">return</span> <span class="nb">id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Type of </span><span class="si">{</span><span class="n">block</span><span class="si">}</span><span class="s2"> not work now&quot;</span><span class="p">)</span>
    <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">block</span><span class="si">}</span><span class="s2"> not found or out of index&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_block_interval" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_block_interval</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">BlockInfo</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">tuple</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Calculate the start(included) and end index (excluded) of the block</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_block_interval</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">block</span><span class="p">:</span> <span class="n">BlockInput</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the start(included) and end index (excluded) of the block&quot;&quot;&quot;</span>
    <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_index</span><span class="p">(</span><span class="n">block</span><span class="p">)</span>
    <span class="n">start</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">index</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">start</span><span class="p">,</span> <span class="n">start</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">length</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_block_length" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_block_length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Return list of blocks</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_block_length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return list of blocks&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_cigar" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_cigar</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target_allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the cigar string of target_allele from ref_allele</p>
<p>If ref_allele not set,
it will automatically find the reference by get_reference</p>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>cigar(list[op_str, num])</code></td>
      <td><p>The list of operator and number of bases</p></td>
    </tr>
  </tbody>
</table>      <p>Exmaple:
  <code>cigar = [(M, 1), (X, 1), (D, 2), (M, 2)]</code></p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_cigar</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">target_allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the cigar string of target_allele from ref_allele</span>

<span class="sd">    If ref_allele not set,</span>
<span class="sd">    it will automatically find the reference by get_reference</span>

<span class="sd">    Return:</span>
<span class="sd">      cigar(list[op_str, num]): The list of operator and number of bases</span>
<span class="sd">    Exmaple:</span>
<span class="sd">      `cigar = [(M, 1), (X, 1), (D, 2), (M, 2)]`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">target_allele</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_allele</span><span class="si">}</span><span class="s2"> not found&quot;</span><span class="p">)</span>
    <span class="n">ref_allele</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_allele_or_error</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">cigar</span><span class="o">.</span><span class="n">calculate_cigar</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">target_allele</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_consensus" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_consensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Generate the consensus sequence by choosing maximum frequency base</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>include_gap</code></td>
        <td><code>bool</code></td>
        <td><p>Allow consensus contains gap if gap is the maximum item.</p>
<p>If include_gap=False and all the base on that position is gap
(not shrinked before),
it will warning and fill with A.</p>
<p><code>E</code> will be ignored.</p></td>
        <td><code>False</code></td>
      </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    
      <div class="highlight"><pre><span></span><code>a0                 CCATT-GGT--GTCGGGTTTCCAG
a1                 CCACTGGGT--ATCGGGTTTCCAG
c2                 CAATTGGGT--GTCGGGT---AAG
consensus          CCATTGGGT--GTCGGGTTTCCAG
consensus(no-gap)  CCATTGGGTAAGTCGGGTTTCCAG
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_consensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">include_gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate the consensus sequence by choosing maximum frequency base</span>

<span class="sd">    Args:</span>
<span class="sd">      include_gap (bool):</span>
<span class="sd">        Allow consensus contains gap if gap is the maximum item.</span>

<span class="sd">        If include_gap=False and all the base on that position is gap</span>
<span class="sd">        (not shrinked before),</span>
<span class="sd">        it will warning and fill with A.</span>

<span class="sd">        `E` will be ignored.</span>

<span class="sd">    Example:</span>
<span class="sd">    ```</span>
<span class="sd">    a0                 CCATT-GGT--GTCGGGTTTCCAG</span>
<span class="sd">    a1                 CCACTGGGT--ATCGGGTTTCCAG</span>
<span class="sd">    c2                 CAATTGGGT--GTCGGGT---AAG</span>
<span class="sd">    consensus          CCATTGGGT--GTCGGGTTTCCAG</span>
<span class="sd">    consensus(no-gap)  CCATTGGGTAAGTCGGGTTTCCAG</span>
<span class="sd">    ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_frequency</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">include_gap</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">4</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;MSA contains gap, try .shrink() before .get_consensus()&quot;</span>
            <span class="p">)</span>
        <span class="n">max_ind</span> <span class="o">=</span> <span class="p">[</span><span class="nb">max</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">max_ind</span> <span class="o">=</span> <span class="p">[</span><span class="nb">max</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">]</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="s2">&quot;ACGT-&quot;</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">max_ind</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_length" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the length of MSA</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the length of MSA&quot;&quot;&quot;</span>
    <span class="c1"># 0 sequences is allow</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
        <span class="c1"># another way to calculate length</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">())</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># when extend, reference allele may not exists here</span>
        <span class="c1"># return len(self.get_reference()[1])</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()))[</span><span class="mi">1</span><span class="p">])</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_reference" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the reference in MSA, if not existed, output the first one</p>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>tuple</code></td>
      <td><p>(allele_name, allele_seq)</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the reference in MSA, if not existed, output the first one</span>

<span class="sd">    Returns:</span>
<span class="sd">      (allele_name, allele_seq)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;MSA is empty&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Reference </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="si">}</span><span class="s2"> not existed in MSA.&quot;</span>
        <span class="s2">&quot; Using the first allele instead.&quot;</span>
    <span class="p">)</span>
    <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="n">allele</span>  <span class="c1"># set it</span>
    <span class="k">return</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_sequence_names" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_sequence_names</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">KeysView</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Same as list_alleles</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_sequence_names</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">KeysView</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Same as list_alleles&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_alleles</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.get_variantion_base" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_variantion_base</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the base positions where variation occurs</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code>msa:
  s0: AAT
  s1: AAC
  s2: CAC
&gt;&gt;&gt; msa.get_variantion_base()
[0, 2]
</code></pre></div>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>positions</code></td>
      <td><p>Each integer represent the position of variation</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_variantion_base</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the base positions where variation occurs</span>

<span class="sd">    Example:</span>
<span class="sd">      ```</span>
<span class="sd">      msa:</span>
<span class="sd">        s0: AAT</span>
<span class="sd">        s1: AAC</span>
<span class="sd">        s2: CAC</span>
<span class="sd">      &gt;&gt;&gt; msa.get_variantion_base()</span>
<span class="sd">      [0, 2]</span>
<span class="sd">      ```</span>

<span class="sd">    Returns:</span>
<span class="sd">      positions:</span>
<span class="sd">        Each integer represent the position of variation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_frequency</span><span class="p">()</span>
    <span class="n">num</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
    <span class="n">base</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">freqs</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">num</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">freq</span><span class="p">:</span>
            <span class="n">base</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">base</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.items" class="doc doc-heading">
<code class="highlight language-python"><span class="n">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>list allele name along with allele sequence like dict.items()</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;list allele name along with allele sequence like dict.items()&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.list_alleles" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_alleles</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">KeysView</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>List all the allele's sequence name in MSA</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">list_alleles</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span>
<span class="p">[</span><span class="s1">&#39;A*01:01:01:01&#39;</span><span class="p">,</span> <span class="s1">&#39;A*01:01:01:02N&#39;</span><span class="p">,</span> <span class="s1">&#39;A*01:01:01:03&#39;</span><span class="p">]</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_alleles</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">KeysView</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    List all the allele&#39;s sequence name in MSA</span>

<span class="sd">    Example:</span>
<span class="sd">       &gt;&gt;&gt; a_gen.list_alleles()[:3]</span>
<span class="sd">       [&#39;A*01:01:01:01&#39;, &#39;A*01:01:01:02N&#39;, &#39;A*01:01:01:03&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.list_blocks" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_blocks</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Return list of blocks</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_blocks</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">BlockInfo</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return list of blocks&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.list_index" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_index</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Return list of index</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_index</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">IndexInfo</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return list of index&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.merge_exon" class="doc doc-heading">
<code class="highlight language-python"><span class="n">merge_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa_nuc</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Merge nuc MSA into gen MSA</p>
<p>It's allow that nuc MSA has new allele name than gen MSA,
Genemsa will add the sequence in MSA, and the intron will fill by <code>E</code></p>
<p>If the exon part of gen MSA is differnet (e.g. less gapped) from nuc MSA,
Genemsa will try to merge if it can</p>
<p>Note that the index will be reset</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code># case1
msa_gen:
  1: &quot;AA|TT|CC&quot;,
  2: &quot;AA|TC|CC&quot;,
msa_nuc:
  3: &quot;TC&quot;,
After merge:
  1: &quot;AA|TT|CC&quot;,
  2: &quot;AA|TC|CC&quot;,
  3: &quot;EE|TC|EE&quot;
</code></pre></div>
<div class="highlight"><pre><span></span><code># case2
msa_gen:
  1: &quot;AA|TT|CC&quot;,
  2: &quot;AA|TC|CC&quot;,
msa_nuc:
  1: &quot;TT-&quot;,
  2: &quot;T-C&quot;,
  4: &quot;TTC&quot;,
After merge:
  1: &quot;AA|TT-|CC&quot;,
  2: &quot;AA|T-C|CC&quot;,
  4: &quot;EE|TTC|EE&quot;
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">merge_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">msa_nuc</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge nuc MSA into gen MSA</span>

<span class="sd">    It&#39;s allow that nuc MSA has new allele name than gen MSA,</span>
<span class="sd">    Genemsa will add the sequence in MSA, and the intron will fill by `E`</span>

<span class="sd">    If the exon part of gen MSA is differnet (e.g. less gapped) from nuc MSA,</span>
<span class="sd">    Genemsa will try to merge if it can</span>

<span class="sd">    Note that the index will be reset</span>

<span class="sd">    Example:</span>
<span class="sd">      ```</span>
<span class="sd">      # case1</span>
<span class="sd">      msa_gen:</span>
<span class="sd">        1: &quot;AA|TT|CC&quot;,</span>
<span class="sd">        2: &quot;AA|TC|CC&quot;,</span>
<span class="sd">      msa_nuc:</span>
<span class="sd">        3: &quot;TC&quot;,</span>
<span class="sd">      After merge:</span>
<span class="sd">        1: &quot;AA|TT|CC&quot;,</span>
<span class="sd">        2: &quot;AA|TC|CC&quot;,</span>
<span class="sd">        3: &quot;EE|TC|EE&quot;</span>
<span class="sd">      ```</span>

<span class="sd">      ```</span>
<span class="sd">      # case2</span>
<span class="sd">      msa_gen:</span>
<span class="sd">        1: &quot;AA|TT|CC&quot;,</span>
<span class="sd">        2: &quot;AA|TC|CC&quot;,</span>
<span class="sd">      msa_nuc:</span>
<span class="sd">        1: &quot;TT-&quot;,</span>
<span class="sd">        2: &quot;T-C&quot;,</span>
<span class="sd">        4: &quot;TTC&quot;,</span>
<span class="sd">      After merge:</span>
<span class="sd">        1: &quot;AA|TT-|CC&quot;,</span>
<span class="sd">        2: &quot;AA|T-C|CC&quot;,</span>
<span class="sd">        4: &quot;EE|TTC|EE&quot;</span>
<span class="sd">      ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># A mapping from gen name to nuc index</span>
    <span class="n">nuc_name_index</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">b</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">())</span> <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span>
    <span class="p">}</span>

    <span class="c1"># check it&#39;s one-to-one mapping</span>
    <span class="n">exon_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()</span> <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">nuc_name_index</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="n">exon_set</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Cannot match blocks: &quot;</span> <span class="sa">f</span><span class="s2">&quot;gen=</span><span class="si">{</span><span class="n">exon_set</span><span class="si">}</span><span class="s2"> nuc=</span><span class="si">{</span><span class="n">nuc_name_index</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="c1"># create new msa and make sure the order of alleles</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">set_blocks</span><span class="p">([])</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">()</span> <span class="o">|</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="c1"># allele names</span>
    <span class="n">gen_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
    <span class="n">nuc_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
    <span class="n">exclude_name</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># block-wise</span>
    <span class="n">msas_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">split_block</span><span class="p">()</span>
    <span class="n">msas_nuc</span> <span class="o">=</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">split_block</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i_gen</span><span class="p">,</span> <span class="n">blk_gen</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">):</span>
        <span class="c1"># intron -&gt; fill with E</span>
        <span class="k">if</span> <span class="n">blk_gen</span><span class="o">.</span><span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">nuc_name_index</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">nuc_names</span> <span class="o">-</span> <span class="n">gen_names</span><span class="p">:</span>
                <span class="n">msas_gen</span><span class="p">[</span><span class="n">i_gen</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;E&quot;</span> <span class="o">*</span> <span class="n">blk_gen</span><span class="o">.</span><span class="n">length</span><span class="p">)</span>
            <span class="n">new_msa</span> <span class="o">+=</span> <span class="n">msas_gen</span><span class="p">[</span><span class="n">i_gen</span><span class="p">]</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span>
                <span class="nb">list</span><span class="p">(</span><span class="n">exclude_name</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
        <span class="c1"># exon -&gt; check before merge</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">i_nuc</span> <span class="o">=</span> <span class="n">nuc_name_index</span><span class="p">[</span><span class="n">blk_gen</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="c1"># content change or length change</span>
            <span class="k">if</span> <span class="n">msas_nuc</span><span class="p">[</span><span class="n">i_nuc</span><span class="p">]</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span> <span class="o">!=</span> <span class="n">msas_gen</span><span class="p">[</span><span class="n">i_gen</span><span class="p">]</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span> <span class="ow">or</span> <span class="nb">any</span><span class="p">(</span>
                <span class="n">msas_nuc</span><span class="p">[</span><span class="n">i_nuc</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">!=</span> <span class="n">msas_gen</span><span class="p">[</span><span class="n">i_gen</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="n">nuc_names</span> <span class="o">&amp;</span> <span class="n">gen_names</span><span class="p">)</span>
            <span class="p">):</span>
                <span class="c1"># check before merge</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gen_names</span> <span class="o">-</span> <span class="n">nuc_names</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Some alleles doesn&#39;t exist in nuc MSA: &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gen_names</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">nuc_names</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

                <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">name</span><span class="p">:</span> <span class="n">msas_nuc</span><span class="p">[</span><span class="n">i_nuc</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="o">!=</span> <span class="n">msas_gen</span><span class="p">[</span><span class="n">i_gen</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
                    <span class="n">gen_names</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">diff_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">diff_names</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Some exon sequences in gen MSA &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;is not same as in nuc MSA &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">blk_gen</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">diff_names</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="n">new_msa</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_names</span><span class="p">)</span>
                    <span class="n">exclude_name</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">diff_names</span><span class="p">)</span>
            <span class="n">new_msa</span> <span class="o">+=</span> <span class="n">msas_nuc</span><span class="p">[</span><span class="n">i_nuc</span><span class="p">]</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">exclude_name</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.meta_to_json" class="doc doc-heading">
<code class="highlight language-python"><span class="n">meta_to_json</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Extract all meta information about this msa into dict(json)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">meta_to_json</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract all meta information about this msa into dict(json)&quot;&quot;&quot;</span>
    <span class="n">meta</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">dataclasses</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">],</span>
        <span class="s2">&quot;blocks&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">dataclasses</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">],</span>
        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_name</span><span class="p">,</span>
        <span class="s2">&quot;reference&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">meta</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.print_alignment" class="doc doc-heading">
<code class="highlight language-python"><span class="n">print_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Print the MSA</p>
<p>Note that the index shown on output header is 1-base absolute position.</p>
<p>Also, the index is absolute index, if you you want continuous positiont,
use <code>.reset_index()</code> first.</p>
<ul>
<li><code>*</code> indicate deletion</li>
<li><code>ATCG</code> indicate the SNV</li>
<li><code>E</code> indicate the null base in exon-only sequence</li>
<li><code>|</code> is intron and exon boundary</li>
</ul>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;A\*.*:01:01:01$&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">])</span>
<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">format_alignment</span><span class="p">()))</span>
<span class="c1"># or simply use</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">a</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">()</span>
                    <span class="mi">3166</span>                                  <span class="mi">3342</span>
                       <span class="o">|</span>                                     <span class="o">|</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGTTAC</span> <span class="n">ACTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGCTAC</span> <span class="n">TCTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGTTAC</span> <span class="n">ACTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGTTAC</span> <span class="n">ACTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">23</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGCTAC</span> <span class="n">TCTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
   <span class="n">A</span><span class="o">*</span><span class="mi">25</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">ATAGAAAAGG</span> <span class="n">AGGGAGCTAC</span> <span class="n">TCTCAGGCTG</span> <span class="n">CAA</span><span class="o">|</span> <span class="n">GCAGTGA</span> <span class="n">CAGTGCCCAG</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">print_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print the MSA</span>

<span class="sd">    Note that the index shown on output header is 1-base absolute position.</span>

<span class="sd">    Also, the index is absolute index, if you you want continuous positiont,</span>
<span class="sd">    use `.reset_index()` first.</span>

<span class="sd">    * `*` indicate deletion</span>
<span class="sd">    * `ATCG` indicate the SNV</span>
<span class="sd">    * `E` indicate the null base in exon-only sequence</span>
<span class="sd">    * `|` is intron and exon boundary</span>

<span class="sd">    Examples:</span>
<span class="sd">      &gt;&gt;&gt; a = msa.select_allele(r&quot;A\\*.*:01:01:01$&quot;).select_exon([6,7])</span>
<span class="sd">      &gt;&gt;&gt; print(&#39;&#39;.join(a.format_alignment()))</span>
<span class="sd">      # or simply use</span>
<span class="sd">      &gt;&gt;&gt; a.print_alignment()</span>
<span class="sd">                          3166                                  3342</span>
<span class="sd">                             |                                     |</span>
<span class="sd">         A*01:01:01:01       ATAGAAAAGG AGGGAGTTAC ACTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">         A*02:01:01:01       ATAGAAAAGG AGGGAGCTAC TCTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">         A*03:01:01:01       ATAGAAAAGG AGGGAGTTAC ACTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">         A*11:01:01:01       ATAGAAAAGG AGGGAGTTAC ACTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">         A*23:01:01:01       ATAGAAAAGG AGGGAGCTAC TCTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">         A*25:01:01:01       ATAGAAAAGG AGGGAGCTAC TCTCAGGCTG CAA| GCAGTGA CAGTGCCCAG</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">page_str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_alignment</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs_format</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">page_str</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.print_alignment_diff" class="doc doc-heading">
<code class="highlight language-python"><span class="n">print_alignment_diff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Print the sequences of all alleles diff from <code>ref_allele</code> sequence.</p>
<p>Same as print_alignment, but
once the base of alleles is same as the base in reference sequence,
<code>-</code> will be used</p>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>str</code></td>
      <td><p>A formatted string</p></td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code>&gt;&gt;&gt; a = msa.select_allele(r&quot;A\*.*:01:01:01$&quot;).select_exon([6,7])
&gt;&gt;&gt; print(a.format_alignment_diff())
                    3166                                  3342
                       |                                     |
   A*01:01:01:01       ATAGAAAAGG AGGGAGTTAC ACTCAGGCTG CAA| GCAGTGA CAGTGCC
   A*02:01:01:01       ---------- ------C--- T--------- ---| ------- -------
   A*03:01:01:01       ---------- ---------- ---------- ---| ------- -------
   A*11:01:01:01       ---------- ---------- ---------- ---| ------- -------
   A*23:01:01:01       ---------- ------C--- T--------- ---| ------- -------
   A*25:01:01:01       ---------- ------C--- T--------- ---| ------- -------
   A*26:01:01:01       ---------- ------C--- T--------- ---| ------- -------
   A*29:01:01:01       ---------- ------C--- T--------- ---| ------- -------
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">print_alignment_diff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print the sequences of all alleles diff from `ref_allele` sequence.</span>

<span class="sd">    Same as print_alignment, but</span>
<span class="sd">    once the base of alleles is same as the base in reference sequence,</span>
<span class="sd">    `-` will be used</span>

<span class="sd">    Returns:</span>
<span class="sd">      str: A formatted string</span>

<span class="sd">    Examples:</span>
<span class="sd">      ```</span>
<span class="sd">      &gt;&gt;&gt; a = msa.select_allele(r&quot;A\\*.*:01:01:01$&quot;).select_exon([6,7])</span>
<span class="sd">      &gt;&gt;&gt; print(a.format_alignment_diff())</span>
<span class="sd">                          3166                                  3342</span>
<span class="sd">                             |                                     |</span>
<span class="sd">         A*01:01:01:01       ATAGAAAAGG AGGGAGTTAC ACTCAGGCTG CAA| GCAGTGA CAGTGCC</span>
<span class="sd">         A*02:01:01:01       ---------- ------C--- T--------- ---| ------- -------</span>
<span class="sd">         A*03:01:01:01       ---------- ---------- ---------- ---| ------- -------</span>
<span class="sd">         A*11:01:01:01       ---------- ---------- ---------- ---| ------- -------</span>
<span class="sd">         A*23:01:01:01       ---------- ------C--- T--------- ---| ------- -------</span>
<span class="sd">         A*25:01:01:01       ---------- ------C--- T--------- ---| ------- -------</span>
<span class="sd">         A*26:01:01:01       ---------- ------C--- T--------- ---| ------- -------</span>
<span class="sd">         A*29:01:01:01       ---------- ------C--- T--------- ---| ------- -------</span>
<span class="sd">      ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calc_diff_msa</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs_format</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.print_alignment_from_center" class="doc doc-heading">
<code class="highlight language-python"><span class="n">print_alignment_from_center</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">pos</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span> <span class="n">left</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">right</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Print all alleles sequences from the center of specific position</p>
<p>Check <code>.format_alignment()</code> for output format detail</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>pos</code></td>
        <td><code>int or list of int</code></td>
        <td><p>The (0-base relative) positions</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>left</code></td>
        <td><code>int</code></td>
        <td><p>How many base shall print at the left of the center</p></td>
        <td><code>5</code></td>
      </tr>
      <tr>
        <td><code>right</code></td>
        <td><code>int</code></td>
        <td><p>How many base shall print at the right of the center</p></td>
        <td><code>5</code></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>str</code></td>
      <td><p>A formatted string</p></td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code> gDNA                    3022
                         |
 HLA-A*03:02        AGAGAAAAAT
 HLA-A*01:40        -----G----
 HLA-A*03:20        -----G----
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">print_alignment_from_center</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span>
    <span class="n">pos</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">]],</span>
    <span class="n">left</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">right</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print all alleles sequences from the center of specific position</span>

<span class="sd">    Check `.format_alignment()` for output format detail</span>

<span class="sd">    Args:</span>
<span class="sd">      pos (int or list of int): The (0-base relative) positions</span>
<span class="sd">      left (int): How many base shall print at the left of the center</span>
<span class="sd">      right (int): How many base shall print at the right of the center</span>

<span class="sd">    Returns:</span>
<span class="sd">      str: A formatted string</span>

<span class="sd">    Examples:</span>
<span class="sd">        ```</span>
<span class="sd">         gDNA                    3022</span>
<span class="sd">                                 |</span>
<span class="sd">         HLA-A*03:02        AGAGAAAAAT</span>
<span class="sd">         HLA-A*01:40        -----G----</span>
<span class="sd">         HLA-A*03:20        -----G----</span>
<span class="sd">        ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
        <span class="n">want_pos</span> <span class="o">=</span> <span class="p">[</span><span class="n">pos</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">want_pos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">want_pos</span><span class="p">:</span>
        <span class="k">return</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calc_center_msa</span><span class="p">(</span><span class="n">want_pos</span><span class="p">,</span> <span class="n">left</span><span class="p">,</span> <span class="n">right</span><span class="p">)</span>
    <span class="n">show_position_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">want_pos</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">(</span><span class="n">show_position_set</span><span class="o">=</span><span class="n">show_position_set</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.print_snv" class="doc doc-heading">
<code class="highlight language-python"><span class="n">print_snv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>A handy function to show all the variation between the alleles</p>
<p>Note: the <code>|</code> in the output string is NOT intron and exon boundary</p>
<p>Check <code>.format_alignment()</code> for output format detail</p>

<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>str</code></td>
      <td><p>A formatted string</p></td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code>Total variantion: 71
                        308    314          329          342    349
                          |      |            |            |      |
 A*01:01:01:01       TGGCCGTCAT GGCGCC| CCCT CCTCCT| ACTC TCGGGGGCCC TGG
 A*02:01:01:01       ---------- ------| ---- -G----| ---- --------T- ---
 A*03:01:01:01       ---------- ------| ---- ------| ---- ---------- ---
 A*11:01:01:01       ---------- ------| ---- ------| ---- ---------- ---
 A*23:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*25:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*26:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*29:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---
 A*30:01:01:01       ---------- ------| ---- ------| ---- ---------- ---
 A*32:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---
 A*33:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---
 A*34:01:01:01       -----A---- ------| ---- -G----| ---- ---------- ---
 A*36:01:01:01       ---------- ------| ---- ------| ---- ---------- ---
 A*66:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*68:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*69:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---
 A*74:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---
 A*80:01:01:01       ---------- -C----| ---- ------| ---- ---------- ---
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">print_snv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A handy function to show all the variation between the alleles</span>

<span class="sd">    Note: the `|` in the output string is NOT intron and exon boundary</span>

<span class="sd">    Check `.format_alignment()` for output format detail</span>

<span class="sd">    Returns:</span>
<span class="sd">      str: A formatted string</span>
<span class="sd">    Example:</span>
<span class="sd">        ```</span>
<span class="sd">        Total variantion: 71</span>
<span class="sd">                                308    314          329          342    349</span>
<span class="sd">                                  |      |            |            |      |</span>
<span class="sd">         A*01:01:01:01       TGGCCGTCAT GGCGCC| CCCT CCTCCT| ACTC TCGGGGGCCC TGG</span>
<span class="sd">         A*02:01:01:01       ---------- ------| ---- -G----| ---- --------T- ---</span>
<span class="sd">         A*03:01:01:01       ---------- ------| ---- ------| ---- ---------- ---</span>
<span class="sd">         A*11:01:01:01       ---------- ------| ---- ------| ---- ---------- ---</span>
<span class="sd">         A*23:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*25:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*26:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*29:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---</span>
<span class="sd">         A*30:01:01:01       ---------- ------| ---- ------| ---- ---------- ---</span>
<span class="sd">         A*32:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---</span>
<span class="sd">         A*33:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---</span>
<span class="sd">         A*34:01:01:01       -----A---- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*36:01:01:01       ---------- ------| ---- ------| ---- ---------- ---</span>
<span class="sd">         A*66:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*68:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*69:01:01:01       ---------- ------| ---- -G----| ---- ---------- ---</span>
<span class="sd">         A*74:01:01:01       ---------- ------| ---- ------| ---- -T-------- ---</span>
<span class="sd">         A*80:01:01:01       ---------- -C----| ---- ------| ---- ---------- ---</span>
<span class="sd">        ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">msa</span><span class="p">,</span> <span class="n">grouped_bases</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calc_snp_msa</span><span class="p">()</span>
    <span class="n">show_position_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="k">for</span> <span class="n">bases</span> <span class="ow">in</span> <span class="n">grouped_bases</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">bases</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;#Total variantion: </span><span class="si">{</span><span class="nb">sum</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">len</span><span class="p">,</span><span class="w"> </span><span class="n">grouped_bases</span><span class="p">))</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">msa</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">_calc_diff_msa</span><span class="p">()</span>
    <span class="n">msa</span><span class="o">.</span><span class="n">print_alignment</span><span class="p">(</span><span class="n">show_position_set</span><span class="o">=</span><span class="n">show_position_set</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs_format</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.remove_allele" class="doc doc-heading">
<code class="highlight language-python"><span class="n">remove_allele</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">inplace</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Remove allele sequences by regex(when query is string)
or by exactly deleting (when query is a list)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">remove_allele</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">inplace</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove allele sequences by regex(when query is string)</span>
<span class="sd">    or by exactly deleting (when query is a list)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
        <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">allele</span><span class="p">):</span>
                <span class="k">del</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">query</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.reset_index" class="doc doc-heading">
<code class="highlight language-python"><span class="n">reset_index</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Reset index:
The old position information will be discard.</p>
<p>Each position information will be counted from 0 and
the label and name will copy from its block information</p>

<p><strong>Examples:</strong></p>
    
      <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">format_alignment_diff</span><span class="p">())</span>
                   <span class="mi">101</span> <span class="mi">103</span> <span class="mi">105</span> <span class="mi">107</span> <span class="mi">109</span> <span class="mi">111</span> <span class="mi">113</span> <span class="mi">115</span> <span class="mi">117</span> <span class="mi">119</span>
                     <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>   <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">G</span>   <span class="n">G</span>   <span class="n">T</span>   <span class="n">C</span>   <span class="n">C</span>   <span class="n">A</span>   <span class="n">C</span>   <span class="n">C</span>   <span class="n">G</span>   <span class="n">A</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>      <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>   <span class="o">-</span>

<span class="o">&gt;&gt;&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">format_alignment_diff</span><span class="p">())</span>
                    <span class="mi">1</span>
                    <span class="o">|</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>      <span class="n">GGTCCACCGA</span>
 <span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">02</span><span class="n">N</span>     <span class="o">----------</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">reset_index</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reset index:</span>
<span class="sd">    The old position information will be discard.</span>

<span class="sd">    Each position information will be counted from 0 and</span>
<span class="sd">    the label and name will copy from its block information</span>

<span class="sd">    Example:</span>
<span class="sd">    ``` python</span>
<span class="sd">    &gt;&gt;&gt; print(a.format_alignment_diff())</span>
<span class="sd">                       101 103 105 107 109 111 113 115 117 119</span>
<span class="sd">                         |   |   |   |   |   |   |   |   |   |</span>
<span class="sd">     A*01:01:01:01       G   G   T   C   C   A   C   C   G   A</span>
<span class="sd">     A*01:01:01:02N      -   -   -   -   -   -   -   -   -   -</span>

<span class="sd">    &gt;&gt;&gt; print(a.reset_index().format_alignment_diff())</span>
<span class="sd">                        1</span>
<span class="sd">                        |</span>
<span class="sd">     A*01:01:01:01      GGTCCACCGA</span>
<span class="sd">     A*01:01:01:02N     ----------</span>
<span class="sd">    ```</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">block</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">length</span><span class="p">):</span>
            <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">IndexInfo</span><span class="p">(</span>
                    <span class="n">pos</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                    <span class="nb">type</span><span class="o">=</span><span class="n">block</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">start</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">assert</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.reverse_complement" class="doc doc-heading">
<code class="highlight language-python"><span class="n">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Reverse the sequences</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reverse the sequences&quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">reversed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">)))</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">reversed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">)))</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
        <span class="p">{</span><span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.save_msa" class="doc doc-heading">
<code class="highlight language-python"><span class="n">save_msa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_fasta</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">file_json</span><span class="p">:</span> <span class="n">FileType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Save Genemsa to pyhlamsa format (fasta and json)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">save_msa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_fasta</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">file_json</span><span class="p">:</span> <span class="n">FileType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save Genemsa to pyhlamsa format (fasta and json)&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="n">file_fasta</span><span class="p">,</span> <span class="n">gap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">f_json</span><span class="p">,</span> <span class="n">require_close</span> <span class="o">=</span> <span class="n">getFileHandle</span><span class="p">(</span><span class="n">file_json</span><span class="p">)</span>
    <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">meta_to_json</span><span class="p">(),</span> <span class="n">f_json</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">require_close</span><span class="p">:</span>
        <span class="n">f_json</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.select_allele" class="doc doc-heading">
<code class="highlight language-python"><span class="n">select_allele</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Select allele name by regex(when query is string)
or by exactly selecting (when query is a list)</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="c1"># select allele name start with A*01:01</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">msa</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;A\*01:01.*&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="c1"># select allele by list of string</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">msa</span><span class="o">.</span><span class="n">select_allele</span><span class="p">([</span><span class="s2">&quot;A*01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:03&quot;</span><span class="p">])</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">select_allele</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Select allele name by regex(when query is string)</span>
<span class="sd">    or by exactly selecting (when query is a list)</span>

<span class="sd">    Examples:</span>
<span class="sd">      &gt;&gt;&gt; # select allele name start with A*01:01</span>
<span class="sd">      &gt;&gt;&gt; msa.select_allele(r&quot;A\\*01:01.*&quot;)</span>
<span class="sd">      &gt;&gt;&gt; # select allele by list of string</span>
<span class="sd">      &gt;&gt;&gt; msa.select_allele([&quot;A*01:01&quot;, &quot;A*02:03&quot;])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="p">{</span><span class="n">allele</span><span class="p">:</span> <span class="n">seq</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">allele</span><span class="p">)}</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">):</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">({</span><span class="n">allele</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span> <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">query</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.select_block" class="doc doc-heading">
<code class="highlight language-python"><span class="n">select_block</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Extract blocks by index</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>index</code></td>
        <td><code>list of int</code></td>
        <td><p>Leave empty if you want all the blocks.</p>
<p>Index start from 0.  e.g.</p>
<ul>
<li>0 for 5-UTR</li>
<li>1 for exon1</li>
<li>2 for intron1</li>
<li>3 for exon2</li>
<li>4 for 3-UTR(for two exons gene)</li>
<li>-1 for last block</li>
</ul>
<p>or you can use list of string,
it will select by block name</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">A</span>  <span class="n">alleles</span><span class="o">=</span><span class="mi">4101</span> <span class="n">block</span><span class="o">=</span><span class="mi">3</span><span class="n">UTR</span><span class="p">(</span><span class="mi">302</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">A</span>  <span class="n">alleles</span><span class="o">=</span><span class="mi">4101</span> <span class="n">block</span><span class="o">=</span><span class="n">intron1</span><span class="p">(</span><span class="mi">130</span><span class="p">)</span> <span class="n">exon2</span><span class="p">(</span><span class="mi">335</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="s2">&quot;5UTR&quot;</span><span class="p">,</span> <span class="s2">&quot;exon3&quot;</span><span class="p">])</span>
<span class="o">&lt;</span><span class="n">A</span>  <span class="n">alleles</span><span class="o">=</span><span class="mi">4101</span> <span class="n">block</span><span class="o">=</span><span class="mi">5</span><span class="n">UTR</span><span class="p">(</span><span class="mi">301</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">413</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">select_block</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">BlockInput</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract blocks by index</span>

<span class="sd">    Args:</span>
<span class="sd">      index (list of int): Leave empty if you want all the blocks.</span>

<span class="sd">        Index start from 0.  e.g.</span>

<span class="sd">        * 0 for 5-UTR</span>
<span class="sd">        * 1 for exon1</span>
<span class="sd">        * 2 for intron1</span>
<span class="sd">        * 3 for exon2</span>
<span class="sd">        * 4 for 3-UTR(for two exons gene)</span>
<span class="sd">        * -1 for last block</span>

<span class="sd">        or you can use list of string,</span>
<span class="sd">        it will select by block name</span>

<span class="sd">    Example:</span>
<span class="sd">      &gt;&gt;&gt; a_gen.select_block([-1])</span>
<span class="sd">      &lt;A  alleles=4101 block=3UTR(302)&gt;</span>

<span class="sd">      &gt;&gt;&gt; a_gen.select_block([2, 3])</span>
<span class="sd">      &lt;A  alleles=4101 block=intron1(130) exon2(335)&gt;</span>

<span class="sd">      &gt;&gt;&gt; a_gen.select_block([&quot;5UTR&quot;, &quot;exon3&quot;])</span>
<span class="sd">      &lt;A  alleles=4101 block=5UTR(301) exon3(413)&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># choose block index by index</span>
    <span class="n">new_block</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">new_index</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_pos</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">index_ids</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">get_block_index</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index_ids</span><span class="p">:</span>
        <span class="n">new_block</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_interval</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">all_pos</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
        <span class="n">new_index</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="n">new_block</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">new_index</span>

    <span class="c1"># extract the sequences inside block region</span>
    <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">gen_seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">new_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">gen_seq</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">all_pos</span><span class="p">])</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">allele</span><span class="p">,</span> <span class="n">new_seq</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.select_complete" class="doc doc-heading">
<code class="highlight language-python"><span class="n">select_complete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Select non exon-only sequences (No <code>E</code> in the sequence)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">select_complete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Select non exon-only sequences (No `E` in the sequence)&quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">({</span><span class="n">allele</span><span class="p">:</span> <span class="n">seq</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="s2">&quot;E&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.select_exon" class="doc doc-heading">
<code class="highlight language-python"><span class="n">select_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">exon_index</span><span class="p">:</span> <span class="n">Iterable</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Extract the exon by index.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>exon_index</code></td>
        <td><code>list[str|int]</code></td>
        <td><p>Index start from 1. i.e.</p>
<ul>
<li>1 for exon1</li>
<li>2 for exon2</li>
</ul>
<p>Leave empty if you want all the exons</p>
<p>If the exon_index contains list of string,
it will select by name</p></td>
        <td><code>[]</code></td>
      </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">2</span><span class="p">]))</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">gen</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7350</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]))</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7350</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">436</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">a_gen</span><span class="o">.</span><span class="n">select_exon</span><span class="p">([</span><span class="s2">&quot;exon2&quot;</span><span class="p">,</span> <span class="s2">&quot;exon3&quot;</span><span class="p">]))</span>
<span class="o">&lt;</span><span class="n">A</span> <span class="n">nuc</span> <span class="n">alleles</span><span class="o">=</span><span class="mi">7350</span> <span class="n">block</span><span class="o">=</span><span class="n">exon2</span><span class="p">(</span><span class="mi">351</span><span class="p">)</span> <span class="n">exon3</span><span class="p">(</span><span class="mi">436</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">select_exon</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">exon_index</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">BlockInput</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the exon by index.</span>

<span class="sd">    Args:</span>
<span class="sd">      exon_index (list[str|int]): Index start from 1. i.e.</span>

<span class="sd">        * 1 for exon1</span>
<span class="sd">        * 2 for exon2</span>

<span class="sd">        Leave empty if you want all the exons</span>

<span class="sd">        If the exon_index contains list of string,</span>
<span class="sd">        it will select by name</span>

<span class="sd">    Example:</span>
<span class="sd">      &gt;&gt;&gt; a_gen.select_exon([2]))</span>
<span class="sd">      &lt;A gen alleles=7350 block=exon2(351)&gt;</span>

<span class="sd">      &gt;&gt;&gt; a_gen.select_exon([2, 3]))</span>
<span class="sd">      &lt;A nuc alleles=7350 block=exon2(351) exon3(436)&gt;</span>

<span class="sd">      &gt;&gt;&gt; a_gen.select_exon([&quot;exon2&quot;, &quot;exon3&quot;]))</span>
<span class="sd">      &lt;A nuc alleles=7350 block=exon2(351) exon3(436)&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exons</span> <span class="o">=</span> <span class="p">[</span><span class="n">b</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()</span> <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">]</span>

    <span class="c1"># If not specific the index, extract all exons</span>
    <span class="n">exon_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">BlockInput</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">exon_index</span><span class="p">:</span>
        <span class="n">exon_list</span> <span class="o">=</span> <span class="n">exons</span>  <span class="c1"># type: ignore</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exon_index</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="c1"># exon -&gt; blocks position</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> is out of exon index&quot;</span><span class="p">)</span>
                <span class="n">i</span> <span class="o">=</span> <span class="n">exons</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">exon_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="c1"># check</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exon_list</span><span class="p">:</span>
        <span class="n">block</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">block</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">block</span><span class="si">}</span><span class="s2"> is not exon: input=</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">select_block</span><span class="p">(</span><span class="n">exon_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.select_incomplete" class="doc doc-heading">
<code class="highlight language-python"><span class="n">select_incomplete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Select exon-only sequences (<code>E</code> exists in the sequence)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">select_incomplete</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Select exon-only sequences (`E` exists in the sequence)&quot;&quot;&quot;</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">extend</span><span class="p">({</span><span class="n">allele</span><span class="p">:</span> <span class="n">seq</span> <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="s2">&quot;E&quot;</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.set_blocks" class="doc doc-heading">
<code class="highlight language-python"><span class="n">set_blocks</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">blocks</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Set blocks (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">set_blocks</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">blocks</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">BlockInfo</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set blocks (inplace)&quot;&quot;&quot;</span>
    <span class="n">new_blocks</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">blocks</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">new_blocks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">BlockInfo</span><span class="p">(</span><span class="n">length</span><span class="o">=</span><span class="n">i</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_blocks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span> <span class="o">!=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">blk</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">blk</span> <span class="ow">in</span> <span class="n">new_blocks</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Total block length not match to alignments&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="n">new_blocks</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.set_reference" class="doc doc-heading">
<code class="highlight language-python"><span class="n">set_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Set the reference in msa (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">set_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">,</span> <span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set the reference in msa (inplace)&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">allele</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot find </span><span class="si">{</span><span class="n">allele</span><span class="si">}</span><span class="s2"> in MSA&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="n">allele</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.shrink" class="doc doc-heading">
<code class="highlight language-python"><span class="n">shrink</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Remove empty base if all bases in that column is gap</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">shrink</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove empty base if all bases in that column is gap&quot;&quot;&quot;</span>
    <span class="c1"># index to delete</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_frequency</span><span class="p">()</span>
    <span class="n">masks</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">!=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">]</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">copy_allele</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># recalcuate blocks</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">)):</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_block_interval</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">blocks</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">masks</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
    <span class="k">assert</span> <span class="nb">sum</span><span class="p">(</span><span class="n">masks</span><span class="p">)</span> <span class="o">==</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">get_length</span><span class="p">()</span>
    <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">masks</span><span class="p">))</span> <span class="k">if</span> <span class="n">masks</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>

    <span class="c1"># remove base in allele</span>
    <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">new_msa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">allele</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span> <span class="k">if</span> <span class="n">masks</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>

    <span class="k">return</span> <span class="n">new_msa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.size" class="doc doc-heading">
<code class="highlight language-python"><span class="n">size</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Get the size (num_of_allele, length_of_sequence)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">size</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the size (num_of_allele, length_of_sequence)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_length</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.sort" class="doc doc-heading">
<code class="highlight language-python"><span class="n">sort</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Sort the allele by their sequences (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">sort</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sort the allele by their sequences (inplace)&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.sort_name" class="doc doc-heading">
<code class="highlight language-python"><span class="n">sort_name</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="o">~</span><span class="n">GenemsaType</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Sort the allele by alelle name (inplace)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">sort_name</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GenemsaType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sort the allele by alelle name (inplace)&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.split_block" class="doc doc-heading">
<code class="highlight language-python"><span class="n">split_block</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="o">~</span><span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Split the msa by blocks</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">split_block</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">GenemsaType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">GenemsaType</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split the msa by blocks&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">list_blocks</span><span class="p">()))]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_MultipleSeqAlignment" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_MultipleSeqAlignment</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MultipleSeqAlignment</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Transfer this object to MultipleSeqAlignment(biopython)</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_MultipleSeqAlignment</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MultipleSeqAlignment</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Transfer this object to MultipleSeqAlignment(biopython)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">MultipleSeqAlignment</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">to_records</span><span class="p">(</span><span class="n">gap</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">annotations</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">meta_to_json</span><span class="p">()</span>
    <span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_bam" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_bam</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fname</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">save_ref</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Save the MSA into bam</p>
<p>All alleles will seen as reads aligned on <code>ref_allele</code></p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>fname</code></td>
        <td><code>str</code></td>
        <td><p>The name of bamfile</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>ref_allele</code></td>
        <td><code>str</code></td>
        <td><p>The reference allele.
  If the ref_allele is empty, the first allele will be reference.</p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>save_ref</code></td>
        <td><code>bool</code></td>
        <td><p>The reference allele will also be saved in the bam file</p></td>
        <td><code>True</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_bam</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fname</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">save_ref</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the MSA into bam</span>

<span class="sd">    All alleles will seen as reads aligned on `ref_allele`</span>

<span class="sd">    Args:</span>
<span class="sd">      fname (str): The name of bamfile</span>
<span class="sd">      ref_allele (str): The reference allele.</span>
<span class="sd">          If the ref_allele is empty, the first allele will be reference.</span>
<span class="sd">      save_ref (bool): The reference allele will also be saved in the bam file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ref_allele</span><span class="p">,</span> <span class="n">ref_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_allele_or_error</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)</span>

    <span class="c1"># setup reference and header</span>
    <span class="n">header</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;HD&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;VN&quot;</span><span class="p">:</span> <span class="s2">&quot;1.0&quot;</span><span class="p">},</span>
        <span class="s2">&quot;SQ&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;LN&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)),</span> <span class="s2">&quot;SN&quot;</span><span class="p">:</span> <span class="n">ref_allele</span><span class="p">}</span>
        <span class="p">],</span>
    <span class="p">}</span>

    <span class="c1"># write bam file</span>
    <span class="k">with</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">header</span><span class="p">)</span> <span class="k">as</span> <span class="n">outf</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># skip</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">save_ref</span> <span class="ow">and</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">ref_allele</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># init bam record</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">()</span>
            <span class="n">a</span><span class="o">.</span><span class="n">query_name</span> <span class="o">=</span> <span class="n">allele</span>
            <span class="n">a</span><span class="o">.</span><span class="n">query_sequence</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">a</span><span class="o">.</span><span class="n">cigar</span> <span class="o">=</span> <span class="n">cigar</span><span class="o">.</span><span class="n">cigar_to_pysam</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_cigar</span><span class="p">(</span><span class="n">allele</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">))</span>  <span class="c1"># type: ignore</span>

            <span class="c1"># set to default</span>
            <span class="n">a</span><span class="o">.</span><span class="n">mapping_quality</span> <span class="o">=</span> <span class="mi">60</span>
            <span class="n">a</span><span class="o">.</span><span class="n">reference_id</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">a</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># a.template_length = 0</span>
            <span class="c1"># a.query_qualities = [30] * len(a.query_sequence)</span>
            <span class="c1"># a.flag = 0</span>
            <span class="c1"># a.next_reference_id = 0</span>
            <span class="c1"># a.next_reference_start = 0</span>
            <span class="n">outf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>

    <span class="n">pysam</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">fname</span><span class="p">,</span> <span class="n">fname</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
    <span class="n">pysam</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_fasta" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fname</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ref_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Save the MSA into fasta</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>gap</code></td>
        <td><code>bool</code></td>
        <td><p>The sequence included gap or not</p></td>
        <td><code>True</code></td>
      </tr>
      <tr>
        <td><code>ref_only</code></td>
        <td><code>bool</code></td>
        <td><p>Save reference sequence only</p></td>
        <td><code>False</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_fasta</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">fname</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ref_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the MSA into fasta</span>

<span class="sd">    Args:</span>
<span class="sd">        gap (bool): The sequence included gap or not</span>
<span class="sd">        ref_only (bool): Save reference sequence only</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ref_only</span><span class="p">:</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">select_allele</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">get_reference</span><span class="p">()[</span><span class="mi">0</span><span class="p">]])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span>
    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">to_records</span><span class="p">(</span><span class="n">gap</span><span class="o">=</span><span class="n">gap</span><span class="p">),</span> <span class="n">fname</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_gff" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fname</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">igv_show_label</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">save_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Save to GFF3 format</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>fname</code></td>
        <td><code>str</code></td>
        <td><p>The file name of gff3</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>strand</code></td>
        <td><code>str</code></td>
        <td><p>Must be "+" or "-".</p>
<p>If the strand is "-", it will add <code>strand</code> in GFF file,
  if you want to reverse the sequence, please use <code>reverse_complement</code> first.</p></td>
        <td><code>&#39;+&#39;</code></td>
      </tr>
      <tr>
        <td><code>ref_allele</code></td>
        <td><code>str</code></td>
        <td><p>The name of allele (Must be the same in save_bam)</p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>igv_show_label</code></td>
        <td><code>bool</code></td>
        <td><p>If it's false, it will generate proper GFF3.
  Set it for True as default for easiler label reading in IGV.</p></td>
        <td><code>False</code></td>
      </tr>
      <tr>
        <td><code>save_all</code></td>
        <td><code>bool</code></td>
        <td><p>Set it True if you want to create gff records for all alleles in msa
  Note that this is not very fast.</p></td>
        <td><code>False</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_gff</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">fname</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span>
    <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span>
    <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">igv_show_label</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">save_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save to GFF3 format</span>

<span class="sd">    Args:</span>
<span class="sd">      fname (str): The file name of gff3</span>
<span class="sd">      strand (str): Must be &quot;+&quot; or &quot;-&quot;.</span>

<span class="sd">          If the strand is &quot;-&quot;, it will add `strand` in GFF file,</span>
<span class="sd">          if you want to reverse the sequence, please use `reverse_complement` first.</span>

<span class="sd">      ref_allele (str): The name of allele (Must be the same in save_bam)</span>
<span class="sd">      igv_show_label (bool): If it&#39;s false, it will generate proper GFF3.</span>
<span class="sd">          Set it for True as default for easiler label reading in IGV.</span>
<span class="sd">      save_all (bool):</span>
<span class="sd">          Set it True if you want to create gff records for all alleles in msa</span>
<span class="sd">          Note that this is not very fast.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># TODO: should I save strand == &#39;-&#39; in model?</span>
    <span class="n">ref_allele</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_allele_or_error</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># labels</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">blocks</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;You should assign block&#39;s label. (We assume seq_type=&#39;other&#39;)&quot;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;other&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">save_all</span><span class="p">:</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">list_alleles</span><span class="p">())</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_allele</span><span class="p">]</span>

    <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">allele</span><span class="si">}</span><span class="s2"> to gff&quot;</span><span class="p">)</span>

        <span class="c1"># remove gap</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">select_allele</span><span class="p">([</span><span class="n">allele</span><span class="p">])</span><span class="o">.</span><span class="n">shrink</span><span class="p">()</span>
        <span class="n">records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">_block_to_gff</span><span class="p">(</span>
                <span class="n">msa</span><span class="o">.</span><span class="n">blocks</span><span class="p">,</span> <span class="n">allele</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="n">igv_show_label</span><span class="o">=</span><span class="n">igv_show_label</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># save</span>
    <span class="n">f_gff</span><span class="p">,</span> <span class="n">require_close</span> <span class="o">=</span> <span class="n">getFileHandle</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
    <span class="n">f_gff</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;##gff-version 3</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">f_gff</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_table_to_string</span><span class="p">(</span><span class="n">records</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">require_close</span><span class="p">:</span>
        <span class="n">f_gff</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_imgt_alignment" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_imgt_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span> <span class="n">seq_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;gen&#39;</span><span class="p">,</span> <span class="n">index_start_from</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;exon1&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Export to IMGT-alignment-like format.</p>
<p>But some features do not implemented:
* AA codon position: So the values in the header always 1
* position: So the values in the header always 1</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>seq_type</code></td>
        <td><code>str</code></td>
        <td><p>gen or nuc format</p></td>
        <td><code>&#39;gen&#39;</code></td>
      </tr>
      <tr>
        <td><code>index_start_from</code></td>
        <td><code>str</code></td>
        <td><p>set the start position (1) starts from.</p></td>
        <td><code>&#39;exon1&#39;</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_imgt_alignment</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">file</span><span class="p">:</span> <span class="n">FileType</span><span class="p">,</span>
    <span class="n">seq_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gen&quot;</span><span class="p">,</span>
    <span class="n">index_start_from</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exon1&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Export to IMGT-alignment-like format.</span>

<span class="sd">    But some features do not implemented:</span>
<span class="sd">    * AA codon position: So the values in the header always 1</span>
<span class="sd">    * position: So the values in the header always 1</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_type: gen or nuc format</span>
<span class="sd">        index_start_from: set the start position (1) starts from.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># This function should be located in io_operation</span>
    <span class="n">new_msa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calc_imgt_alignment_msa</span><span class="p">()</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">get_block_interval</span><span class="p">(</span><span class="n">index_start_from</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># The position is calculated by reference sequence not MSA itself</span>
    <span class="n">ref_no_gap_pos</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">get_reference</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ind</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
            <span class="n">ref_no_gap_pos</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">ref_no_gap_pos</span>

    <span class="k">if</span> <span class="n">seq_type</span> <span class="o">==</span> <span class="s2">&quot;gen&quot;</span><span class="p">:</span>
        <span class="c1"># force modify the index</span>
        <span class="c1"># IMGT ends with -1 and start with 1</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">pos</span> <span class="o">-=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">pos</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">ind</span><span class="o">.</span><span class="n">pos</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="n">pages</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_format_page</span><span class="p">(</span>
            <span class="p">[</span><span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">],</span>
            <span class="n">header_text_align</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
            <span class="n">index_header</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gDNA&quot;</span><span class="p">],</span>
            <span class="n">show_position_when_same_value</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">seq_type</span> <span class="o">==</span> <span class="s2">&quot;nuc&quot;</span><span class="p">:</span>
        <span class="n">fake_index</span> <span class="o">=</span> <span class="p">[</span><span class="n">IndexInfo</span><span class="p">(</span><span class="n">pos</span><span class="o">=</span><span class="mi">0</span><span class="p">)]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">pages</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_format_page</span><span class="p">(</span>
            <span class="p">[</span><span class="n">new_msa</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">fake_index</span><span class="p">],</span>
            <span class="n">header_text_align</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
            <span class="n">index_header</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cDNA&quot;</span><span class="p">,</span> <span class="s2">&quot;AA Codon&quot;</span><span class="p">],</span>
            <span class="n">show_position_when_same_value</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Not implement </span><span class="si">{</span><span class="n">seq_type</span><span class="si">=}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># write</span>
    <span class="n">txt</span><span class="p">,</span> <span class="n">require_close</span> <span class="o">=</span> <span class="n">getFileHandle</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">txt</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">new_msa</span><span class="o">.</span><span class="n">_apply_page</span><span class="p">,</span> <span class="n">pages</span><span class="p">)))</span>
    <span class="k">if</span> <span class="n">require_close</span><span class="p">:</span>
        <span class="n">txt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_records" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_records</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Transfer MSA to list of SeqRecord.</p>
<p>If the base is <code>E</code>
which indicate the bases are not existed for exon-only sequence,
those beses are treated as gap.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>gap</code></td>
        <td><code>bool</code></td>
        <td><p>The sequence included gap or not</p></td>
        <td><code>True</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_records</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transfer MSA to list of SeqRecord.</span>

<span class="sd">    If the base is `E`</span>
<span class="sd">    which indicate the bases are not existed for exon-only sequence,</span>
<span class="sd">    those beses are treated as gap.</span>

<span class="sd">    Args:</span>
<span class="sd">      gap (bool): The sequence included gap or not</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">gap</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)),</span> <span class="nb">id</span><span class="o">=</span><span class="n">allele</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="n">SeqRecord</span><span class="p">(</span>
                <span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)),</span>
                <span class="nb">id</span><span class="o">=</span><span class="n">allele</span><span class="p">,</span>
                <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">allele</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.to_vcf" class="doc doc-heading">
<code class="highlight language-python"><span class="n">to_vcf</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_vcf</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">save_ref</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">plain_text</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>Save Genemsa into vcf format</p>
<p>It will save msa into sorted and normalized vcf file (xxx.vcf.gz)
along with vcf's index (xxx.vcf.gz.tbi)
(You can still output plain vcf without any manipulation by set plain_text=True)</p>
<p>Note that vcf-format discard the per-base alignment especially
when there are many snps/indels mixed together in that position.
Also, after vcf is normalized, the msa may not be the same MSA alignment as before.
(But sequences are still the same)</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>ref_allele</code></td>
        <td><code>str</code></td>
        <td><p>The name of reference allele.
I recommend the reference allele contains no gaps.</p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>save_ref</code></td>
        <td><code>bool</code></td>
        <td><p>The reference allele will also be saved in the vcf file</p></td>
        <td><code>True</code></td>
      </tr>
      <tr>
        <td><code>plain_text</code></td>
        <td><code>bool</code></td>
        <td><p>Disable sort vcf and index vcf</p></td>
        <td><code>False</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">to_vcf</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">file_vcf</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ref_allele</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">save_ref</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">plain_text</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save Genemsa into vcf format</span>

<span class="sd">    It will save msa into sorted and normalized vcf file (xxx.vcf.gz)</span>
<span class="sd">    along with vcf&#39;s index (xxx.vcf.gz.tbi)</span>
<span class="sd">    (You can still output plain vcf without any manipulation by set plain_text=True)</span>

<span class="sd">    Note that vcf-format discard the per-base alignment especially</span>
<span class="sd">    when there are many snps/indels mixed together in that position.</span>
<span class="sd">    Also, after vcf is normalized, the msa may not be the same MSA alignment as before.</span>
<span class="sd">    (But sequences are still the same)</span>

<span class="sd">    Args:</span>
<span class="sd">      ref_allele (str):</span>
<span class="sd">        The name of reference allele.</span>
<span class="sd">        I recommend the reference allele contains no gaps.</span>
<span class="sd">      save_ref (bool): The reference allele will also be saved in the vcf file</span>
<span class="sd">      plain_text (bool): Disable sort vcf and index vcf</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ref_allele</span><span class="p">,</span> <span class="n">ref_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_allele_or_error</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">)</span>

    <span class="c1"># extract all variants from all alleles</span>
    <span class="n">allele_variants</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_alleles</span><span class="p">():</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">save_ref</span> <span class="ow">and</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">ref_allele</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">extract_variants</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">allele</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
            <span class="n">v</span><span class="o">.</span><span class="n">chrom</span> <span class="o">=</span> <span class="n">ref_allele</span>
        <span class="n">allele_variants</span><span class="p">[</span><span class="n">allele</span><span class="p">]</span> <span class="o">=</span> <span class="n">variants</span>

    <span class="c1"># remove suffix</span>
    <span class="k">if</span> <span class="n">file_vcf</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf.gz&quot;</span><span class="p">):</span>
        <span class="n">basename</span> <span class="o">=</span> <span class="n">file_vcf</span><span class="p">[:</span><span class="o">-</span><span class="mi">7</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">file_vcf</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">):</span>
        <span class="n">basename</span> <span class="o">=</span> <span class="n">file_vcf</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">basename</span> <span class="o">=</span> <span class="n">file_vcf</span>
    <span class="n">tmpname</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.tmp&quot;</span>

    <span class="c1"># to vcf</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_vcf</span><span class="p">:</span>
        <span class="n">f_vcf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">get_vcf_header</span><span class="p">(</span><span class="n">ref_allele</span><span class="p">,</span> <span class="n">ref_seq</span><span class="p">))</span>
        <span class="n">f_vcf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">f_vcf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_table_to_string</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">variants_to_table</span><span class="p">(</span><span class="n">allele_variants</span><span class="p">)))</span>
    <span class="k">if</span> <span class="n">plain_text</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.vcf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.vcf&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># sort, normalize and index</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">select_allele</span><span class="p">([</span><span class="n">ref_allele</span><span class="p">])</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.fa&quot;</span><span class="p">,</span> <span class="n">gap</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.norm.vcf.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_vcf</span><span class="p">:</span>
        <span class="n">f_vcf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="n">bcftools</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span>  <span class="c1"># type: ignore</span>
                <span class="s2">&quot;-f&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.fa&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;-O&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.vcf.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_vcf</span><span class="p">:</span>
        <span class="n">f_vcf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bcftools</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.norm.vcf.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;-O&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">))</span>  <span class="c1"># type: ignore</span>
    <span class="n">bcftools</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.vcf.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="s2">&quot;-f&quot;</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.fa&quot;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.fa.fai&quot;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.vcf&quot;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tmpname</span><span class="si">}</span><span class="s2">.norm.vcf.gz&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene.genemsa.Genemsa.truth" class="doc doc-heading">
<code class="highlight language-python"><span class="n">truth</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>If msa has 0 alleles return False
else return Implement if(self) function</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene/genemsa.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">truth</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    If msa has 0 alleles return False</span>
<span class="sd">    else return Implement if(self) function</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">bool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">

<a id="pyhlamsa.gene_family.hla"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h2 id="pyhlamsa.gene_family.hla.HLAmsa" class="doc doc-heading">
        <code>
HLAmsa            (<span title="pyhlamsa.gene_family.family.Familymsa">Familymsa</span>)
        </code>



</h2>

    <div class="doc doc-contents ">

      <p>A HLA interface.</p>
<p>This module read the HLA MSA from alignments/*.txt files,
which use <code>|</code> to separate the intron/exon/UTR regions, but
the labels are assumed with the order:
<code>5UTR exon1 intron1 exon2 ... exonN 3UTR</code></p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>dict[str, Genemsa]</code></td>
        <td><p>The dictionary use gene_name as key and msa object as value</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">HLAmsa</span><span class="p">(</span><span class="n">Familymsa</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A HLA interface.</span>

<span class="sd">    This module read the HLA MSA from alignments/*.txt files,</span>
<span class="sd">    which use `|` to separate the intron/exon/UTR regions, but</span>
<span class="sd">    the labels are assumed with the order:</span>
<span class="sd">    `5UTR exon1 intron1 exon2 ... exonN 3UTR`</span>

<span class="sd">    Attributes:</span>
<span class="sd">      genes (dict[str, Genemsa]):</span>
<span class="sd">        The dictionary use gene_name as key and msa object as value</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
        <span class="n">imgt_alignment_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">                Set None if you want read all gene in HLA</span>

<span class="sd">            filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">                If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">            imgt_alignment_folder (str): Path to your IMGT-formatted MSA alignment file</span>

<span class="sd">                You can manually download the folder from</span>
<span class="sd">                &lt;https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments&gt;</span>
<span class="sd">                or &lt;http://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/&gt;</span>
<span class="sd">                and unzip Alignments_Rel_3470.zip</span>

<span class="sd">                Otherwise it will automatically download the database to</span>
<span class="sd">                `imgt_alignment_folder`. Default is `./alignment_v{verion}`</span>

<span class="sd">            version (str): IMGT version you want to download (e.g. 3470 for 3.47.0)</span>

<span class="sd">                If `imgt_alignment_folder` is existed, this value will be ignored.</span>
<span class="sd">                Use `Latest` to get latest version.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">imgt_alignment_folder</span><span class="p">:</span>
            <span class="n">imgt_alignment_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;alignment_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">imgt_alignment_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_download_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Download the IMGTHLA alignments folder to `db_folder`</span>

<span class="sd">        I get the alignments folder from git instead of</span>
<span class="sd">        &lt;http://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/&gt;</span>
<span class="sd">        for better version controlling</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmp_dir</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span>
                <span class="s2">&quot;git&quot;</span><span class="p">,</span>
                <span class="s2">&quot;clone&quot;</span><span class="p">,</span>
                <span class="s2">&quot;--branch&quot;</span><span class="p">,</span>
                <span class="n">version</span><span class="p">,</span>
                <span class="s2">&quot;--single-branch&quot;</span><span class="p">,</span>
                <span class="s2">&quot;https://github.com/ANHIG/IMGTHLA.git&quot;</span><span class="p">,</span>
                <span class="n">tmp_dir</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span><span class="s2">&quot;mv&quot;</span><span class="p">,</span> <span class="n">tmp_dir</span> <span class="o">+</span> <span class="s2">&quot;/alignments&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">search_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Handy function to list names from file pattern&quot;&quot;&quot;</span>
        <span class="n">arr_files</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="n">search_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">set</span><span class="p">([</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">arr_files</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
        <span class="n">drb</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;DRB1&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB3&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB4&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB5&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/*_gen.txt&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/*_nuc.txt&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span>
            <span class="c1"># * E exon7 is ambiguous, exon8 is gone (Also exon8 is not pseudo exon)</span>
            <span class="c1">#    &lt;E gen alleles=258 block=5UTR(301) exon1(64) intron1(130)</span>
            <span class="c1">#        exon2(270) intron2(244) exon3(276) intron3(621) exon4(276)</span>
            <span class="c1">#        intron4(124) exon5(117) intron5(751) exon6(33) intron6(104)</span>
            <span class="c1">#        exon7(43) intron7(165) exon8(5) 3UTR(356)&gt;</span>
            <span class="c1">#    &lt;E nuc alleles=262 block=exon1(64) exon2(270) exon3(276)</span>
            <span class="c1">#        exon4(276) exon5(117) exon6(33) exon7(41)&gt;</span>
            <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
                <span class="n">names</span> <span class="o">=</span> <span class="n">names</span> <span class="o">-</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;E&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read `{gene}_{filetype}.txt`.</span>

<span class="sd">        If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.txt&quot;</span><span class="p">)</span>
            <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="o">!=</span> <span class="s2">&quot;P&quot;</span><span class="p">:</span>
                <span class="c1"># P: special case: has even block</span>
                <span class="c1"># &lt;P gen alleles=5 block=(475) (261) (589) (276) (124) (117)</span>
                <span class="c1">#                        (412) (33) (150) (48) (163) (297)&gt;</span>
                <span class="n">msa_gen</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gen </span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># Special Case: DRB* nuc are in DRB_nuc.txt</span>
            <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;DRB&quot;</span><span class="p">):</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/DRB_nuc.txt&quot;</span><span class="p">)</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;.*&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.txt&quot;</span>
                <span class="p">)</span>

            <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="n">msa_nuc</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nuc </span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># remove some gene</span>
            <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Remove alleles doesn&#39;t exist in gen and nuc either: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

            <span class="c1"># merge</span>
            <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged </span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">msa_merged</span>

        <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_gen</span>
        <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_nuc</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.__getitem__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get specific gene&#39;s msa&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gen&#39;</span><span class="p">,</span> <span class="s1">&#39;nuc&#39;</span><span class="p">],</span> <span class="n">imgt_alignment_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;Latest&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>str | list[str]</code></td>
        <td><p>A list of genes you want to read.</p>
<p>Set None if you want read all gene in HLA</p></td>
        <td><code>None</code></td>
      </tr>
      <tr>
        <td><code>filetype</code></td>
        <td><code>str | list[str] | set[str]</code></td>
        <td><p>A list of filetype.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them automatically.</p></td>
        <td><code>[&#39;gen&#39;, &#39;nuc&#39;]</code></td>
      </tr>
      <tr>
        <td><code>imgt_alignment_folder</code></td>
        <td><code>str</code></td>
        <td><p>Path to your IMGT-formatted MSA alignment file</p>
<p>You can manually download the folder from
<a href="https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments">https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments</a>
or <a href="http://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/">http://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/</a>
and unzip Alignments_Rel_3470.zip</p>
<p>Otherwise it will automatically download the database to
<code>imgt_alignment_folder</code>. Default is <code>./alignment_v{verion}</code></p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>version</code></td>
        <td><code>str</code></td>
        <td><p>IMGT version you want to download (e.g. 3470 for 3.47.0)</p>
<p>If <code>imgt_alignment_folder</code> is existed, this value will be ignored.
Use <code>Latest</code> to get latest version.</p></td>
        <td><code>&#39;Latest&#39;</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
    <span class="n">imgt_alignment_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">            Set None if you want read all gene in HLA</span>

<span class="sd">        filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">            If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">        imgt_alignment_folder (str): Path to your IMGT-formatted MSA alignment file</span>

<span class="sd">            You can manually download the folder from</span>
<span class="sd">            &lt;https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments&gt;</span>
<span class="sd">            or &lt;http://ftp.ebi.ac.uk/pub/databases/ipd/imgt/hla/&gt;</span>
<span class="sd">            and unzip Alignments_Rel_3470.zip</span>

<span class="sd">            Otherwise it will automatically download the database to</span>
<span class="sd">            `imgt_alignment_folder`. Default is `./alignment_v{verion}`</span>

<span class="sd">        version (str): IMGT version you want to download (e.g. 3470 for 3.47.0)</span>

<span class="sd">            If `imgt_alignment_folder` is existed, this value will be ignored.</span>
<span class="sd">            Use `Latest` to get latest version.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">imgt_alignment_folder</span><span class="p">:</span>
        <span class="n">imgt_alignment_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;alignment_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
        <span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">imgt_alignment_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span>
    <span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.__iter__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iter gene name like iter(dict)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.items" class="doc doc-heading">
<code class="highlight language-python"><span class="n">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">genemsa</span><span class="o">.</span><span class="n">Genemsa</span><span class="p">]]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>list gene name and msa like dict.items()</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Genemsa</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;list gene name and msa like dict.items()&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.list_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>List the gene in folder</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
    <span class="n">drb</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;DRB1&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB3&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB4&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB5&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/*_gen.txt&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/*_nuc.txt&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span>
        <span class="c1"># * E exon7 is ambiguous, exon8 is gone (Also exon8 is not pseudo exon)</span>
        <span class="c1">#    &lt;E gen alleles=258 block=5UTR(301) exon1(64) intron1(130)</span>
        <span class="c1">#        exon2(270) intron2(244) exon3(276) intron3(621) exon4(276)</span>
        <span class="c1">#        intron4(124) exon5(117) intron5(751) exon6(33) intron6(104)</span>
        <span class="c1">#        exon7(43) intron7(165) exon8(5) 3UTR(356)&gt;</span>
        <span class="c1">#    &lt;E nuc alleles=262 block=exon1(64) exon2(270) exon3(276)</span>
        <span class="c1">#        exon4(276) exon5(117) exon6(33) exon7(41)&gt;</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names</span> <span class="o">-</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;E&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.list_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>List all the gene's name in this family</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List all the gene&#39;s name in this family&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla.HLAmsa.read_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>Read <code>{gene}_{filetype}.txt</code>.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them.</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read `{gene}_{filetype}.txt`.</span>

<span class="sd">    If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.txt&quot;</span><span class="p">)</span>
        <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="o">!=</span> <span class="s2">&quot;P&quot;</span><span class="p">:</span>
            <span class="c1"># P: special case: has even block</span>
            <span class="c1"># &lt;P gen alleles=5 block=(475) (261) (589) (276) (124) (117)</span>
            <span class="c1">#                        (412) (33) (150) (48) (163) (297)&gt;</span>
            <span class="n">msa_gen</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gen </span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># Special Case: DRB* nuc are in DRB_nuc.txt</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;DRB&quot;</span><span class="p">):</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/DRB_nuc.txt&quot;</span><span class="p">)</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;.*&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_imgt_alignment</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.txt&quot;</span>
            <span class="p">)</span>

        <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="n">msa_nuc</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nuc </span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># remove some gene</span>
        <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Remove alleles doesn&#39;t exist in gen and nuc either: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

        <span class="c1"># merge</span>
        <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged </span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">msa_merged</span>

    <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_gen</span>
    <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_nuc</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">

<a id="pyhlamsa.gene_family.hla_ex"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h2 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX" class="doc doc-heading">
        <code>
HLAmsaEX            (<span title="pyhlamsa.gene_family.family.Familymsa">Familymsa</span>)
        </code>



</h2>

    <div class="doc doc-contents ">

      <p>A HLA interface but read gene from MSF and
read intron/exon information from hla.dat</p>
<p>I think this one is more reliable</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>dict[str, Genemsa]</code></td>
        <td><p>The dictionary use gene_name as key and msa object as value</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">HLAmsaEX</span><span class="p">(</span><span class="n">Familymsa</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A HLA interface but read gene from MSF and</span>
<span class="sd">    read intron/exon information from hla.dat</span>

<span class="sd">    I think this one is more reliable</span>

<span class="sd">    Attributes:</span>
<span class="sd">      genes (dict[str, Genemsa]):</span>
<span class="sd">        The dictionary use gene_name as key and msa object as value</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
        <span class="n">imgt_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">                Set None if you want read all gene in HLA</span>

<span class="sd">            filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">                If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">            imgt_folder (str): Path to your IMGT/HLA root folder</span>

<span class="sd">                You can manually download &lt;https://github.com/ANHIG/IMGTHLA/&gt; and</span>
<span class="sd">                checkout to specific branch</span>

<span class="sd">                Or it will automatically download the database with assigned `version` to</span>
<span class="sd">                `imgt_folder`. Default is `./IMGT_v{verion}`</span>

<span class="sd">            version (str): IMGT version you want to download (e.g. 3470 for 3.47.0)</span>

<span class="sd">                If `imgt_folder` is existed, this value will be ignored.</span>

<span class="sd">                You can use `Latest` to get the latest version</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">imgt_folder</span><span class="p">:</span>
            <span class="n">imgt_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IMGT_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">imgt_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_download_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Download the IMGT/HLA msf and hla.dat to folder `imgt_folder`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span><span class="s2">&quot;git&quot;</span><span class="p">,</span> <span class="s2">&quot;lfs&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;git lfs not installed (Try apt install git-lfs)&quot;</span><span class="p">)</span>
            <span class="n">exit</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span>
            <span class="s2">&quot;git&quot;</span><span class="p">,</span>
            <span class="s2">&quot;clone&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--branch&quot;</span><span class="p">,</span>
            <span class="n">version</span><span class="p">,</span>
            <span class="s2">&quot;--single-branch&quot;</span><span class="p">,</span>
            <span class="s2">&quot;https://github.com/ANHIG/IMGTHLA.git&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span><span class="s2">&quot;git&quot;</span><span class="p">,</span> <span class="s2">&quot;lfs&quot;</span><span class="p">,</span> <span class="s2">&quot;pull&quot;</span><span class="p">,</span> <span class="n">cwd</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">search_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract name from file pattern&quot;&quot;&quot;</span>
        <span class="n">arr_files</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="n">search_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">set</span><span class="p">([</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">arr_files</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
        <span class="n">drb</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;DRB1&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB3&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB4&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB5&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_gen.msf&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_nuc.msf&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># HLA-E nuc has 7 bases differences from exon7 and exon8</span>
            <span class="n">names</span> <span class="o">=</span> <span class="p">(</span><span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read `msf/{gene}_{filetype}.msf` and hla.dat.</span>

<span class="sd">        If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;dat&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reading hla.dat&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/hla.dat&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.msf&quot;</span><span class="p">)</span>
            <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># Special Case: DRB* nuc are in DRB_nuc.txt</span>
            <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;DRB&quot;</span><span class="p">):</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/DRB_nuc.msf&quot;</span><span class="p">)</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;.*&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.msf&quot;</span><span class="p">)</span>
            <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># remove some gen not included in nuc</span>
            <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Remove alleles existed in gen but not in nuc: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

            <span class="c1"># merge</span>
            <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">msa_merged</span>
        <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_gen</span>
        <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_nuc</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.__getitem__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get specific gene&#39;s msa&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gen&#39;</span><span class="p">,</span> <span class="s1">&#39;nuc&#39;</span><span class="p">],</span> <span class="n">imgt_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;Latest&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>str | list[str]</code></td>
        <td><p>A list of genes you want to read.</p>
<p>Set None if you want read all gene in HLA</p></td>
        <td><code>None</code></td>
      </tr>
      <tr>
        <td><code>filetype</code></td>
        <td><code>str | list[str] | set[str]</code></td>
        <td><p>A list of filetype.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them automatically.</p></td>
        <td><code>[&#39;gen&#39;, &#39;nuc&#39;]</code></td>
      </tr>
      <tr>
        <td><code>imgt_folder</code></td>
        <td><code>str</code></td>
        <td><p>Path to your IMGT/HLA root folder</p>
<p>You can manually download <a href="https://github.com/ANHIG/IMGTHLA/">https://github.com/ANHIG/IMGTHLA/</a> and
checkout to specific branch</p>
<p>Or it will automatically download the database with assigned <code>version</code> to
<code>imgt_folder</code>. Default is <code>./IMGT_v{verion}</code></p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>version</code></td>
        <td><code>str</code></td>
        <td><p>IMGT version you want to download (e.g. 3470 for 3.47.0)</p>
<p>If <code>imgt_folder</code> is existed, this value will be ignored.</p>
<p>You can use <code>Latest</code> to get the latest version</p></td>
        <td><code>&#39;Latest&#39;</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
    <span class="n">imgt_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">            Set None if you want read all gene in HLA</span>

<span class="sd">        filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">            If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">        imgt_folder (str): Path to your IMGT/HLA root folder</span>

<span class="sd">            You can manually download &lt;https://github.com/ANHIG/IMGTHLA/&gt; and</span>
<span class="sd">            checkout to specific branch</span>

<span class="sd">            Or it will automatically download the database with assigned `version` to</span>
<span class="sd">            `imgt_folder`. Default is `./IMGT_v{verion}`</span>

<span class="sd">        version (str): IMGT version you want to download (e.g. 3470 for 3.47.0)</span>

<span class="sd">            If `imgt_folder` is existed, this value will be ignored.</span>

<span class="sd">            You can use `Latest` to get the latest version</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">imgt_folder</span><span class="p">:</span>
        <span class="n">imgt_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IMGT_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">imgt_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.__iter__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iter gene name like iter(dict)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.items" class="doc doc-heading">
<code class="highlight language-python"><span class="n">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">genemsa</span><span class="o">.</span><span class="n">Genemsa</span><span class="p">]]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>list gene name and msa like dict.items()</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Genemsa</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;list gene name and msa like dict.items()&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.list_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>List the gene in folder</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
    <span class="n">drb</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;DRB1&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB3&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB4&quot;</span><span class="p">,</span> <span class="s2">&quot;DRB5&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_gen.msf&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_nuc.msf&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="n">drb</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># HLA-E nuc has 7 bases differences from exon7 and exon8</span>
        <span class="n">names</span> <span class="o">=</span> <span class="p">(</span><span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.list_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>List all the gene's name in this family</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List all the gene&#39;s name in this family&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.hla_ex.HLAmsaEX.read_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>Read <code>msf/{gene}_{filetype}.msf</code> and hla.dat.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them.</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/hla_ex.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read `msf/{gene}_{filetype}.msf` and hla.dat.</span>

<span class="sd">    If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;dat&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reading hla.dat&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/hla.dat&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.msf&quot;</span><span class="p">)</span>
        <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># Special Case: DRB* nuc are in DRB_nuc.txt</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;DRB&quot;</span><span class="p">):</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/DRB_nuc.msf&quot;</span><span class="p">)</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_allele</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;.*&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.msf&quot;</span><span class="p">)</span>
        <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># remove some gen not included in nuc</span>
        <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Remove alleles existed in gen but not in nuc: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

        <span class="c1"># merge</span>
        <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">msa_merged</span>
    <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_gen</span>
    <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_nuc</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">

<a id="pyhlamsa.gene_family.kir"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h2 id="pyhlamsa.gene_family.kir.KIRmsa" class="doc doc-heading">
        <code>
KIRmsa            (<span title="pyhlamsa.gene_family.family.Familymsa">Familymsa</span>)
        </code>



</h2>

    <div class="doc doc-contents ">

      <p>A KIR interface that read MSA from MSF and KIR.dat</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>dict[str, Genemsa]</code></td>
        <td><p>The msa object for each gene</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">KIRmsa</span><span class="p">(</span><span class="n">Familymsa</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A KIR interface that read MSA from MSF and KIR.dat</span>

<span class="sd">    Attributes:</span>
<span class="sd">        genes (dict[str, Genemsa]): The msa object for each gene</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
        <span class="n">ipd_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">                Set None if you want read all gene in HLA</span>

<span class="sd">            filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">                If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">            ipd_folder (str): Path to your IPD/KIR folder</span>

<span class="sd">                You can manually download &lt;https://github.com/ANHIG/IPDKIR&gt; and</span>
<span class="sd">                checkout to specific branch</span>

<span class="sd">                Or it will automatically download the database with assigned `version` to</span>
<span class="sd">                `ipd_folder`. Default is `./kIR_v{verion}`</span>

<span class="sd">            version (str): IMGT version you want to download (2110 for version 2.11.0)</span>

<span class="sd">                If `ipd_folder` is existed, this value will be ignored.</span>

<span class="sd">                version=&#39;Latest&#39; to get latest version, however sometime it cannot work</span>
<span class="sd">                because database may change the format, or contains bugs</span>
<span class="sd">                (e.g. 2.11.0 https://github.com/ANHIG/IPDKIR/issues/44)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Why not version 2110 -&gt; 2DL4,2DL5 has exon 4</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ipd_folder</span><span class="p">:</span>
            <span class="n">ipd_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;KIR_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">ipd_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_download_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Download the KIR to `IPDKIR`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_run_shell</span><span class="p">(</span>
            <span class="s2">&quot;git&quot;</span><span class="p">,</span>
            <span class="s2">&quot;clone&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--branch&quot;</span><span class="p">,</span>
            <span class="n">version</span><span class="p">,</span>
            <span class="s2">&quot;--single-branch&quot;</span><span class="p">,</span>
            <span class="s2">&quot;https://github.com/ANHIG/IPDKIR&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">search_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract name from file pattern&quot;&quot;&quot;</span>
        <span class="n">arr_files</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="n">search_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">set</span><span class="p">([</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">arr_files</span><span class="p">])</span>

    <span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_gen.msf&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># Most&#39;s of E nuc is different from dat record</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_nuc.msf&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read `{gene}_{filetype}.msf and kir.dat`.</span>

<span class="sd">        If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;dat&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reading kir.dat&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/KIR.dat&quot;</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/KIR.dat&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/kir.dat&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.msf&quot;</span><span class="p">)</span>
            <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gen </span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.msf&quot;</span><span class="p">)</span>
            <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nuc </span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="c1"># remove some gen not included in nuc</span>
            <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Remove alleles doesn&#39;t exist in gen and nuc either: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
            <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

            <span class="c1"># specical case</span>
            <span class="c1"># exon 3 is pseudo exon</span>
            <span class="c1"># so, fill with gene&#39;s exon3</span>
            <span class="n">gene_has_pseudo_exon3</span> <span class="o">=</span> <span class="p">[</span>
                <span class="s2">&quot;KIR2DL1&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DL2&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DL3&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DP1&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DS1&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DS2&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DS3&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DS4&quot;</span><span class="p">,</span>
                <span class="s2">&quot;KIR2DS5&quot;</span><span class="p">,</span>
            <span class="p">]</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_has_pseudo_exon3</span><span class="p">:</span>
                <span class="n">exon3</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="mi">5</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">):</span>
                    <span class="n">exon3</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="n">exon3</span><span class="o">.</span><span class="n">get_length</span><span class="p">())</span>
                <span class="n">msa_nuc</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_block</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)))</span>
                    <span class="o">+</span> <span class="n">exon3</span>
                    <span class="o">+</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_block</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">blocks</span><span class="p">))))</span>
                <span class="p">)</span>

            <span class="c1"># merge</span>
            <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged </span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">msa_merged</span>

        <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_gen</span>
        <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">msa_nuc</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.__getitem__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get specific gene&#39;s msa&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gen&#39;</span><span class="p">,</span> <span class="s1">&#39;nuc&#39;</span><span class="p">],</span> <span class="n">ipd_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;Latest&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>str | list[str]</code></td>
        <td><p>A list of genes you want to read.</p>
<p>Set None if you want read all gene in HLA</p></td>
        <td><code>None</code></td>
      </tr>
      <tr>
        <td><code>filetype</code></td>
        <td><code>str | list[str] | set[str]</code></td>
        <td><p>A list of filetype.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them automatically.</p></td>
        <td><code>[&#39;gen&#39;, &#39;nuc&#39;]</code></td>
      </tr>
      <tr>
        <td><code>ipd_folder</code></td>
        <td><code>str</code></td>
        <td><p>Path to your IPD/KIR folder</p>
<p>You can manually download <a href="https://github.com/ANHIG/IPDKIR">https://github.com/ANHIG/IPDKIR</a> and
checkout to specific branch</p>
<p>Or it will automatically download the database with assigned <code>version</code> to
<code>ipd_folder</code>. Default is <code>./kIR_v{verion}</code></p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>version</code></td>
        <td><code>str</code></td>
        <td><p>IMGT version you want to download (2110 for version 2.11.0)</p>
<p>If <code>ipd_folder</code> is existed, this value will be ignored.</p>
<p>version='Latest' to get latest version, however sometime it cannot work
because database may change the format, or contains bugs
(e.g. 2.11.0 https://github.com/ANHIG/IPDKIR/issues/44)</p></td>
        <td><code>&#39;Latest&#39;</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gen&quot;</span><span class="p">,</span> <span class="s2">&quot;nuc&quot;</span><span class="p">],</span>
    <span class="n">ipd_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Latest&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">            Set None if you want read all gene in HLA</span>

<span class="sd">        filetype (str | list[str] | set[str]): A list of filetype.</span>

<span class="sd">            If both `gen` and `nuc` are given, it will merge them automatically.</span>

<span class="sd">        ipd_folder (str): Path to your IPD/KIR folder</span>

<span class="sd">            You can manually download &lt;https://github.com/ANHIG/IPDKIR&gt; and</span>
<span class="sd">            checkout to specific branch</span>

<span class="sd">            Or it will automatically download the database with assigned `version` to</span>
<span class="sd">            `ipd_folder`. Default is `./kIR_v{verion}`</span>

<span class="sd">        version (str): IMGT version you want to download (2110 for version 2.11.0)</span>

<span class="sd">            If `ipd_folder` is existed, this value will be ignored.</span>

<span class="sd">            version=&#39;Latest&#39; to get latest version, however sometime it cannot work</span>
<span class="sd">            because database may change the format, or contains bugs</span>
<span class="sd">            (e.g. 2.11.0 https://github.com/ANHIG/IPDKIR/issues/44)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Why not version 2110 -&gt; 2DL4,2DL5 has exon 4</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">ipd_folder</span><span class="p">:</span>
        <span class="n">ipd_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;KIR_v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">ipd_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.__iter__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iter gene name like iter(dict)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.items" class="doc doc-heading">
<code class="highlight language-python"><span class="n">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">genemsa</span><span class="o">.</span><span class="n">Genemsa</span><span class="p">]]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>list gene name and msa like dict.items()</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Genemsa</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;list gene name and msa like dict.items()&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.list_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>List the gene in folder</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_gen.msf&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># Most&#39;s of E nuc is different from dat record</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_nuc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_name</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/*_nuc.msf&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="n">names_gen</span> <span class="o">&amp;</span> <span class="n">names_nuc</span>
    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.list_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>List all the gene's name in this family</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List all the gene&#39;s name in this family&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.kir.KIRmsa.read_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>Read <code>{gene}_{filetype}.msf and kir.dat</code>.</p>
<p>If both <code>gen</code> and <code>nuc</code> are given, it will merge them.</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/kir.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read `{gene}_{filetype}.msf and kir.dat`.</span>

<span class="sd">    If both `gen` and `nuc` are given, it will merge them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;dat&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reading kir.dat&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/KIR.dat&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/KIR.dat&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">read_dat_block</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/kir.dat&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_gen.msf&quot;</span><span class="p">)</span>
        <span class="n">msa_gen</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;gen&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gen </span><span class="si">{</span><span class="n">msa_gen</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="o">.</span><span class="n">read_msf_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/msf/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_nuc.msf&quot;</span><span class="p">)</span>
        <span class="n">msa_nuc</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="n">msa_nuc</span> <span class="o">=</span> <span class="n">dat</span><span class="o">.</span><span class="n">apply_dat_info_on_msa</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dat</span><span class="p">,</span> <span class="n">seq_type</span><span class="o">=</span><span class="s2">&quot;nuc&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nuc </span><span class="si">{</span><span class="n">msa_nuc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span> <span class="ow">and</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="c1"># remove some gen not included in nuc</span>
        <span class="n">diff_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">get_sequence_names</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Remove alleles doesn&#39;t exist in gen and nuc either: </span><span class="si">{</span><span class="n">diff_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="n">msa_gen</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">remove_allele</span><span class="p">(</span><span class="n">diff_name</span><span class="p">)</span>

        <span class="c1"># specical case</span>
        <span class="c1"># exon 3 is pseudo exon</span>
        <span class="c1"># so, fill with gene&#39;s exon3</span>
        <span class="n">gene_has_pseudo_exon3</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;KIR2DL1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DL2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DL3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DP1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DS1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DS2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DS3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DS4&quot;</span><span class="p">,</span>
            <span class="s2">&quot;KIR2DS5&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_has_pseudo_exon3</span><span class="p">:</span>
            <span class="n">exon3</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">select_block</span><span class="p">([</span><span class="mi">5</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">msa_gen</span><span class="p">):</span>
                <span class="n">exon3</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="n">exon3</span><span class="o">.</span><span class="n">get_length</span><span class="p">())</span>
            <span class="n">msa_nuc</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_block</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)))</span>
                <span class="o">+</span> <span class="n">exon3</span>
                <span class="o">+</span> <span class="n">msa_nuc</span><span class="o">.</span><span class="n">select_block</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">msa_nuc</span><span class="o">.</span><span class="n">blocks</span><span class="p">))))</span>
            <span class="p">)</span>

        <span class="c1"># merge</span>
        <span class="n">msa_merged</span> <span class="o">=</span> <span class="n">msa_gen</span><span class="o">.</span><span class="n">merge_exon</span><span class="p">(</span><span class="n">msa_nuc</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged </span><span class="si">{</span><span class="n">msa_merged</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">msa_merged</span>

    <span class="k">elif</span> <span class="s2">&quot;gen&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_gen</span>
    <span class="k">elif</span> <span class="s2">&quot;nuc&quot;</span> <span class="ow">in</span> <span class="n">filetype</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">msa_nuc</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gen or nuc are not exist in filetype&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">

<a id="pyhlamsa.gene_family.cyp"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h2 id="pyhlamsa.gene_family.cyp.CYPmsa" class="doc doc-heading">
        <code>
CYPmsa            (<span title="pyhlamsa.gene_family.family.Familymsa">Familymsa</span>)
        </code>



</h2>

    <div class="doc doc-contents ">

      <p>A CYP interface that read MSA from fasta and haplotypes.tsv</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>dict[str, Genemsa]</code></td>
        <td><p>The msa object for each gene</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">CYPmsa</span><span class="p">(</span><span class="n">Familymsa</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A CYP interface that read MSA from fasta and haplotypes.tsv</span>

<span class="sd">    Attributes:</span>
<span class="sd">        genes (dict[str, Genemsa]): The msa object for each gene</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[],</span>
        <span class="n">pharmvar_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;5.2.2&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">                Set None if you want read all gene in HLA</span>

<span class="sd">            filetype: Ignore</span>

<span class="sd">            pharmvar_folder (str): Path to your pharmvar folder</span>

<span class="sd">                You should manually download &lt;https://www.pharmvar.org/download&gt;</span>
<span class="sd">                and unzip it.</span>

<span class="sd">            version (str): PharmVar version you want to download</span>

<span class="sd">                Not works now</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pharmvar_folder</span><span class="p">:</span>
            <span class="n">pharmvar_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;pharmvar-</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">pharmvar_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_download_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Download the CYP from https://www.pharmvar.org/download</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;You should download CYP genes from https://www.pharmvar.org/download&quot;</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read `{gene}/{gene}.haplotypes.fasta` and `{gene}/{gene}.haplotypes.tsv`</span>

<span class="sd">        `filetype` will be ignored now</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># read fasta</span>
        <span class="n">ref_seqs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.haplotypes.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span>
        <span class="p">):</span>
            <span class="c1"># split allele name</span>
            <span class="c1"># e.g.  rs75017182, rs56038477 PV01077 NG_008807.2 PharmVar Version:5.1.10</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
                <span class="n">ref_seqs</span><span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read sequence </span><span class="si">{</span><span class="n">ref_seqs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># read tsv</span>
        <span class="c1"># split by &#39;\t&#39; and ignore header</span>
        <span class="n">var_text</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">/RefSeqGene/*.haplotypes.tsv&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">var_text</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">)][</span><span class="mi">1</span><span class="p">:]</span>

        <span class="c1"># Get Reference sequence</span>
        <span class="n">reference_seq</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">reference_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ref_seqs</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;REFERENCE&quot;</span> <span class="ow">and</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="mi">0</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">reference_seq</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="n">reference_seq</span> <span class="o">==</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                    <span class="k">continue</span>
                <span class="n">reference_name</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">reference_seq</span> <span class="o">=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_seq</span><span class="p">:</span>
            <span class="c1"># hard-coded for DPYD, because it&#39;s reference `NG_008807.2` is missing</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;DPYD&quot;</span><span class="p">:</span>
                <span class="n">reference_name</span> <span class="o">=</span> <span class="s2">&quot;rs111858276&quot;</span>
                <span class="n">reference_seq</span> <span class="o">=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]</span>
                <span class="n">reference_seq</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">reference_seq</span><span class="p">[:</span> <span class="mi">376460</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;A&quot;</span> <span class="o">+</span> <span class="n">reference_seq</span><span class="p">[</span><span class="mi">376460</span><span class="p">:]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Not reference found in </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">assert</span> <span class="n">reference_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="c1"># Fill with Reference and reference is the first one</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="p">{</span><span class="n">reference_name</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
        <span class="n">alleles</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span><span class="n">allele_name</span><span class="p">:</span> <span class="n">reference_seq</span> <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">table</span><span class="p">)}</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read vcf </span><span class="si">{</span><span class="n">alleles</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Remove duplciate</span>
        <span class="c1"># in 5.1.10, there exist two same row</span>
        <span class="c1"># CYP2D6*149  CYP2D6  rs59421388  NG_008376.4 8203    8203    G   A   substitution</span>
        <span class="c1"># CYP2D6*149  CYP2D6  rs59421388  NG_008376.4 8203    8203    G   A   substitution</span>
        <span class="n">table_unique1</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;REFERENCE&quot;</span><span class="p">,</span> <span class="n">table</span><span class="p">)</span>
        <span class="n">table_unique2</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="nb">map</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="n">table_unique1</span><span class="p">))</span>  <span class="c1"># type: ignore</span>
        <span class="n">table_unique</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">map</span><span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">table_unique2</span><span class="p">))))</span>  <span class="c1"># type: ignore</span>

        <span class="c1"># Reconstruct msa from VCF variant</span>
        <span class="c1"># VCF type: substitution and deleteion, insertion</span>
        <span class="c1"># Table Header: [&#39;Haplotype Name&#39;, &#39;Gene&#39;, &#39;rsID&#39;, &#39;ReferenceSequence&#39;,</span>
        <span class="c1">#   &#39;Variant Start&#39;, &#39;Variant Stop&#39;, &#39;Reference Allele&#39;, &#39;Variant Allele&#39;, &#39;Type&#39;]</span>
        <span class="c1"># Insertion will move the index, so insert from last position first</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">table_unique</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])):</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;substitution&quot;</span><span class="p">:</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;deletion&quot;</span><span class="p">:</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
                <span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
            <span class="k">elif</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;insertion&quot;</span><span class="p">:</span>
                <span class="c1"># hard coded bugs</span>
                <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;CYP2A6*31&quot;</span><span class="p">):</span>
                    <span class="c1"># CYP2A6*31.001   4530    4552    CCCCTTCCTGAGACCCTTAACCC -   deletion</span>
                    <span class="c1"># CYP2A6*31.001   4530    4531    -   AATCCATATGTGGAATCTG insertion</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;CYP2A6*27&quot;</span><span class="p">):</span>
                    <span class="c1"># CYP2A6*27.001   7183    7184    -   A   insertion</span>
                    <span class="k">continue</span>

                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
                <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
                <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
                <span class="c1"># check the gap is already inserted by others</span>
                <span class="c1"># if false: insert &#39;-&#39; for all alleles</span>
                <span class="c1"># else: replace the &#39;-&#39; with insertion seq</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">pos</span><span class="p">:</span><span class="n">end</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">):</span>
                    <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
                        <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">][:</span><span class="n">pos</span><span class="p">]</span>
                            <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
                            <span class="o">+</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">][</span><span class="n">pos</span><span class="p">:]</span>
                        <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown Type </span><span class="si">{</span><span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">assert</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">pos</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
            <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][:</span><span class="n">pos</span><span class="p">]</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">+</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">end</span><span class="p">:]</span>

        <span class="c1"># split allele name (alleles.key() will change)</span>
        <span class="c1"># e.g. rs75017182, rs56038477  DPYD    rs75017182  NG_008807.2 346167</span>
        <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">alleles</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="k">if</span> <span class="s2">&quot;, &quot;</span> <span class="ow">in</span> <span class="n">allele_name</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">an</span> <span class="ow">in</span> <span class="n">allele_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
                    <span class="n">alleles</span><span class="p">[</span><span class="n">an</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="o">=</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span>
                <span class="k">del</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span>

        <span class="c1"># double check</span>
        <span class="n">length</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">length</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">allele_name</span><span class="si">}</span><span class="s2"> is not same as reference&quot;</span><span class="p">)</span>

        <span class="k">assert</span> <span class="n">length</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">msa</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="n">alleles</span>
        <span class="n">msa</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="n">BlockInfo</span><span class="p">(</span><span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">msa</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;other&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.__getitem__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get specific gene&#39;s msa&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">pharmvar_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;5.2.2&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>genes</code></td>
        <td><code>str | list[str]</code></td>
        <td><p>A list of genes you want to read.</p>
<p>Set None if you want read all gene in HLA</p></td>
        <td><code>None</code></td>
      </tr>
      <tr>
        <td><code>filetype</code></td>
        <td><code>Union[str, Iterable[str]]</code></td>
        <td><p>Ignore</p></td>
        <td><code>[]</code></td>
      </tr>
      <tr>
        <td><code>pharmvar_folder</code></td>
        <td><code>str</code></td>
        <td><p>Path to your pharmvar folder</p>
<p>You should manually download <a href="https://www.pharmvar.org/download">https://www.pharmvar.org/download</a>
and unzip it.</p></td>
        <td><code>&#39;&#39;</code></td>
      </tr>
      <tr>
        <td><code>version</code></td>
        <td><code>str</code></td>
        <td><p>PharmVar version you want to download</p>
<p>Not works now</p></td>
        <td><code>&#39;5.2.2&#39;</code></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">GeneSet</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">pharmvar_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;5.2.2&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        genes (str | list[str]): A list of genes you want to read.</span>

<span class="sd">            Set None if you want read all gene in HLA</span>

<span class="sd">        filetype: Ignore</span>

<span class="sd">        pharmvar_folder (str): Path to your pharmvar folder</span>

<span class="sd">            You should manually download &lt;https://www.pharmvar.org/download&gt;</span>
<span class="sd">            and unzip it.</span>

<span class="sd">        version (str): PharmVar version you want to download</span>

<span class="sd">            Not works now</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">pharmvar_folder</span><span class="p">:</span>
        <span class="n">pharmvar_folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;pharmvar-</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">filetype</span><span class="p">,</span> <span class="n">db_folder</span><span class="o">=</span><span class="n">pharmvar_folder</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.__iter__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">


        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iter gene name like iter(dict)&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.items" class="doc doc-heading">
<code class="highlight language-python"><span class="n">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pyhlamsa</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">genemsa</span><span class="o">.</span><span class="n">Genemsa</span><span class="p">]]</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>list gene name and msa like dict.items()</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Genemsa</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;list gene name and msa like dict.items()&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.list_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>List the gene in folder</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List the gene in folder&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.list_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-inherited"><code>inherited</code></small>
  </span>

</h3>

    <div class="doc doc-contents ">

      <p>List all the gene's name in this family</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">list_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List all the gene&#39;s name in this family&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h3 id="pyhlamsa.gene_family.cyp.CYPmsa.read_db_gene" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span></code>


</h3>

    <div class="doc doc-contents ">

      <p>Read <code>{gene}/{gene}.haplotypes.fasta</code> and <code>{gene}/{gene}.haplotypes.tsv</code></p>
<p><code>filetype</code> will be ignored now</p>

        <details class="quote">
          <summary>Source code in <code>pyhlamsa/gene_family/cyp.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_db_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">filetype</span><span class="p">:</span> <span class="n">TypeSet</span> <span class="o">=</span> <span class="p">[])</span> <span class="o">-&gt;</span> <span class="n">Genemsa</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read `{gene}/{gene}.haplotypes.fasta` and `{gene}/{gene}.haplotypes.tsv`</span>

<span class="sd">    `filetype` will be ignored now</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># read fasta</span>
    <span class="n">ref_seqs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.haplotypes.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span>
    <span class="p">):</span>
        <span class="c1"># split allele name</span>
        <span class="c1"># e.g.  rs75017182, rs56038477 PV01077 NG_008807.2 PharmVar Version:5.1.10</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
            <span class="n">ref_seqs</span><span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read sequence </span><span class="si">{</span><span class="n">ref_seqs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># read tsv</span>
    <span class="c1"># split by &#39;\t&#39; and ignore header</span>
    <span class="n">var_text</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">db_folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">/RefSeqGene/*.haplotypes.tsv&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">table</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">var_text</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">)][</span><span class="mi">1</span><span class="p">:]</span>

    <span class="c1"># Get Reference sequence</span>
    <span class="n">reference_seq</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">reference_name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ref_seqs</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;REFERENCE&quot;</span> <span class="ow">and</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">reference_seq</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">reference_seq</span> <span class="o">==</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                <span class="k">continue</span>
            <span class="n">reference_name</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">reference_seq</span> <span class="o">=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_seq</span><span class="p">:</span>
        <span class="c1"># hard-coded for DPYD, because it&#39;s reference `NG_008807.2` is missing</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s2">&quot;DPYD&quot;</span><span class="p">:</span>
            <span class="n">reference_name</span> <span class="o">=</span> <span class="s2">&quot;rs111858276&quot;</span>
            <span class="n">reference_seq</span> <span class="o">=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]</span>
            <span class="n">reference_seq</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">reference_seq</span><span class="p">[:</span> <span class="mi">376460</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;A&quot;</span> <span class="o">+</span> <span class="n">reference_seq</span><span class="p">[</span><span class="mi">376460</span><span class="p">:]</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Not reference found in </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">assert</span> <span class="n">reference_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="c1"># Fill with Reference and reference is the first one</span>
    <span class="n">alleles</span> <span class="o">=</span> <span class="p">{</span><span class="n">reference_name</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
    <span class="n">alleles</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span><span class="n">allele_name</span><span class="p">:</span> <span class="n">reference_seq</span> <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">table</span><span class="p">)}</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Read vcf </span><span class="si">{</span><span class="n">alleles</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Remove duplciate</span>
    <span class="c1"># in 5.1.10, there exist two same row</span>
    <span class="c1"># CYP2D6*149  CYP2D6  rs59421388  NG_008376.4 8203    8203    G   A   substitution</span>
    <span class="c1"># CYP2D6*149  CYP2D6  rs59421388  NG_008376.4 8203    8203    G   A   substitution</span>
    <span class="n">table_unique1</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;REFERENCE&quot;</span><span class="p">,</span> <span class="n">table</span><span class="p">)</span>
    <span class="n">table_unique2</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="nb">map</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="n">table_unique1</span><span class="p">))</span>  <span class="c1"># type: ignore</span>
    <span class="n">table_unique</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">map</span><span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">table_unique2</span><span class="p">))))</span>  <span class="c1"># type: ignore</span>

    <span class="c1"># Reconstruct msa from VCF variant</span>
    <span class="c1"># VCF type: substitution and deleteion, insertion</span>
    <span class="c1"># Table Header: [&#39;Haplotype Name&#39;, &#39;Gene&#39;, &#39;rsID&#39;, &#39;ReferenceSequence&#39;,</span>
    <span class="c1">#   &#39;Variant Start&#39;, &#39;Variant Stop&#39;, &#39;Reference Allele&#39;, &#39;Variant Allele&#39;, &#39;Type&#39;]</span>
    <span class="c1"># Insertion will move the index, so insert from last position first</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">table_unique</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])):</span>
        <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;substitution&quot;</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;deletion&quot;</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
            <span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
        <span class="k">elif</span> <span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;insertion&quot;</span><span class="p">:</span>
            <span class="c1"># hard coded bugs</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;CYP2A6*31&quot;</span><span class="p">):</span>
                <span class="c1"># CYP2A6*31.001   4530    4552    CCCCTTCCTGAGACCCTTAACCC -   deletion</span>
                <span class="c1"># CYP2A6*31.001   4530    4531    -   AATCCATATGTGGAATCTG insertion</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;CYP2A6*27&quot;</span><span class="p">):</span>
                <span class="c1"># CYP2A6*27.001   7183    7184    -   A   insertion</span>
                <span class="k">continue</span>

            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
            <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
            <span class="c1"># check the gap is already inserted by others</span>
            <span class="c1"># if false: insert &#39;-&#39; for all alleles</span>
            <span class="c1"># else: replace the &#39;-&#39; with insertion seq</span>
            <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">pos</span><span class="p">:</span><span class="n">end</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
                    <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">][:</span><span class="n">pos</span><span class="p">]</span>
                        <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
                        <span class="o">+</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">][</span><span class="n">pos</span><span class="p">:]</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown Type </span><span class="si">{</span><span class="n">i</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">assert</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">pos</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
        <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][:</span><span class="n">pos</span><span class="p">]</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">+</span> <span class="n">alleles</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">end</span><span class="p">:]</span>

    <span class="c1"># split allele name (alleles.key() will change)</span>
    <span class="c1"># e.g. rs75017182, rs56038477  DPYD    rs75017182  NG_008807.2 346167</span>
    <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">alleles</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="k">if</span> <span class="s2">&quot;, &quot;</span> <span class="ow">in</span> <span class="n">allele_name</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">an</span> <span class="ow">in</span> <span class="n">allele_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
                <span class="n">alleles</span><span class="p">[</span><span class="n">an</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="o">=</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span>
            <span class="k">del</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span>

    <span class="c1"># double check</span>
    <span class="n">length</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">allele_name</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">length</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">alleles</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ref_seqs</span><span class="p">[</span><span class="n">allele_name</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">allele_name</span><span class="si">}</span><span class="s2"> is not same as reference&quot;</span><span class="p">)</span>

    <span class="k">assert</span> <span class="n">length</span>
    <span class="n">msa</span> <span class="o">=</span> <span class="n">Genemsa</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">msa</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="n">alleles</span>
    <span class="n">msa</span><span class="o">.</span><span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="n">BlockInfo</span><span class="p">(</span><span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">msa</span><span class="o">.</span><span class="n">assume_label</span><span class="p">(</span><span class="s2">&quot;other&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": ["toc.integrate"], "search": "assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.d7c377c4.min.js"></script>
      
    
  </body>
</html>